<!DOCTYPE html>
<html lang="en-us">

<head>

  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="generator" content="Source Themes Academic 4.5.0">

  

  
  
  
  
  
    
    
    
  
  

  <meta name="author" content="Cecina Babich Morrow">

  
  
  
    
  
  <meta name="description" content="">

  
  <link rel="alternate" hreflang="en-us" href="/post/">

  


  
  
  
  <meta name="theme-color" content="hsl(339, 90%, 68%)">
  

  
  
  
  
    
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/academicons/1.8.6/css/academicons.min.css" integrity="sha256-uFVgMKfistnJAfoCUQigIl+JfUaP47GrRKjf6CTPVmw=" crossorigin="anonymous">
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/all.min.css" integrity="sha256-+N4/V/SbAFiW1MPBCXnfnP9QSN3+Keu+NlB+0ev/YKQ=" crossorigin="anonymous">
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/fancybox/3.5.7/jquery.fancybox.min.css" integrity="sha256-Vzbj7sDDS/woiFS3uNKo8eIuni59rjyNGtXfstRzStA=" crossorigin="anonymous">

    
    
    
      
    
    
      
      
        
          <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.15.10/styles/github.min.css" crossorigin="anonymous" title="hl-light" disabled>
          <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.15.10/styles/dracula.min.css" crossorigin="anonymous" title="hl-dark">
        
      
    

    

    

  

  
  
  
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Montserrat:400,700%7CRoboto:400,400italic,700%7CRoboto+Mono&display=swap">
  

  
  
  
  
  <link rel="stylesheet" href="/css/academic.css">

  




  


  

  <link rel="manifest" href="/index.webmanifest">
  <link rel="icon" type="image/png" href="/img/icon-32.png">
  <link rel="apple-touch-icon" type="image/png" href="/img/icon-192.png">

  <link rel="canonical" href="/post/">

  
  
  
  
    
    
  
  
  <meta property="twitter:card" content="summary">
  
  <meta property="twitter:site" content="@babichmorrowc">
  <meta property="twitter:creator" content="@babichmorrowc">
  
  <meta property="og:site_name" content="Cecina Babich Morrow">
  <meta property="og:url" content="/post/">
  <meta property="og:title" content="Posts | Cecina Babich Morrow">
  <meta property="og:description" content=""><meta property="og:image" content="/img/icon-192.png">
  <meta property="twitter:image" content="/img/icon-192.png"><meta property="og:locale" content="en-us">
  
    
      <meta property="article:published_time" content="2019-10-22T00:00:00&#43;00:00">
    
    <meta property="article:modified_time" content="2019-10-22T00:00:00&#43;00:00">
  

  



  


  


  





  <title>Posts | Cecina Babich Morrow</title>

</head>

<body id="top" data-spy="scroll" data-offset="70" data-target="#navbar-main" class="dark">

  <aside class="search-results" id="search">
  <div class="container">
    <section class="search-header">

      <div class="row no-gutters justify-content-between mb-3">
        <div class="col-6">
          <h1>Search</h1>
        </div>
        <div class="col-6 col-search-close">
          <a class="js-search" href="#"><i class="fas fa-times-circle text-muted" aria-hidden="true"></i></a>
        </div>
      </div>

      <div id="search-box">
        
        <input name="q" id="search-query" placeholder="Search..." autocapitalize="off"
        autocomplete="off" autocorrect="off" spellcheck="false" type="search">
        
      </div>

    </section>
    <section class="section-search-results">

      <div id="search-hits">
        
      </div>

    </section>
  </div>
</aside>


  
<nav class="navbar navbar-light fixed-top navbar-expand-lg py-0 compensate-for-scrollbar" id="navbar-main">
  <div class="container">

    
      <a class="navbar-brand" href="/">Cecina Babich Morrow</a>
      
      <button type="button" class="navbar-toggler" data-toggle="collapse"
              data-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
        <span><i class="fas fa-bars"></i></span>
      </button>
      

    
    <div class="collapse navbar-collapse" id="navbar">

      
      
      <ul class="navbar-nav mr-auto">
        

        

        
        
        
          
        

        
        
        
        
        
        
          
          
          
            
          
          
        

        <li class="nav-item">
          <a class="nav-link " href="/#about"><span>About</span></a>
        </li>

        
        

        

        
        
        
          
        

        
        
        
        
        
        
          
          
          
            
          
          
        

        <li class="nav-item">
          <a class="nav-link " href="/#publications"><span>Publications</span></a>
        </li>

        
        

        

        
        
        
          
        

        
        
        
        
        
        
          
          
          
            
          
          
        

        <li class="nav-item">
          <a class="nav-link " href="/#projects"><span>Projects</span></a>
        </li>

        
        

        

        
        
        
          
        

        
        
        
        
        
        
          
          
          
            
          
          
        

        <li class="nav-item">
          <a class="nav-link " href="/#posts"><span>Posts</span></a>
        </li>

        
        

        

        
        
        
          
        

        
        
        
        
        
        
          
          
          
            
          
          
        

        <li class="nav-item">
          <a class="nav-link " href="/#contact"><span>Contact</span></a>
        </li>

        
        

      
      </ul>
      <ul class="navbar-nav ml-auto">
      

        

        
        <li class="nav-item">
          <a class="nav-link js-search" href="#"><i class="fas fa-search" aria-hidden="true"></i></a>
        </li>
        

        

        
        <li class="nav-item">
          <a class="nav-link js-dark-toggle" href="#"><i class="fas fa-moon" aria-hidden="true"></i></a>
        </li>
        

      </ul>

    </div>
  </div>
</nav>


  
<span class="js-widget-page d-none"></span>




  







  
  
  

  

  

  

  

  

  
  

  
  
  

  
  
  
  
  

  
  

  <section id="2019-10-11-google_fonts" class="home-section wg-blank   "  >
    <div class="container">
      


<div class="row">
  
    <div class="col-12 col-lg-4 section-heading">
      <h1>Putting Google Fonts in R graphs</h1>
      
    </div>
    <div class="col-12 col-lg-8">
      


<figure>
<img class="special-img-class" style="width:60%" src="/img/lego_statue.jpg" />
<figcaption>
In keeping with my Lego theme, this statue from Nathan Sawaya’s exhibition at the New York Hall of Science represents some of the frustrations of adding fonts to R.
</figcaption>
</figure>
<div id="inspiration-for-this-post" class="section level2">
<h2>Inspiration for this post</h2>
<p>I’ve been working on creating a visualization cookbook for R graphics (check out <a href="https://bbc.github.io/rcookbook/">this great example</a> from the BBC to see what I’m eventually going for). As part of this process, I wanted to be able to change the font of my plots to <a href="https://fonts.google.com/specimen/Source+Sans+Pro">Source Sans Pro</a>, which is a Google font.</p>
</div>
<div id="sysfonts-package" class="section level2">
<h2><code>sysfonts</code> package</h2>
<p>The first step here was to install and load the <code>sysfonts</code> package:</p>
<pre class="r"><code>library(sysfonts)</code></pre>
<div id="xquartz" class="section level3">
<h3>XQuartz</h3>
<p>If you end up getting the following error when loading <code>sysfonts</code>: <code>Reason: image not found</code>, you might need to do some additional work. This happened to me initially, and after doing some Googling, I found <a href="https://github.com/yixuan/showtext/issues/13">this GitHub issue</a>. Installing XQuartz was recommended as a fix, particularly for Macs, so I downloaded it from <a href="https://www.xquartz.org/">here</a>. After downloading XQuartz, I uninstalled, reinstalled, and loaded <code>sysfonts</code> and things went smoothly.</p>
</div>
</div>
<div id="downloading-font-to-your-computer" class="section level2">
<h2>Downloading font to your computer</h2>
<p>As I learned from this <a href="https://github.com/kjhealy/sourcesans/blob/150207b18f4fbcc4c954715f269f8b20ce834a54/R/sourcesans.r">super helpful GitHub repo</a>, you need to install the font on your system as well as within R.</p>
<div id="mac" class="section level3">
<h3>Mac</h3>
<p>I went through this process on a Mac, so this is what I ended up doing to install the font. I went to the <a href="https://fonts.google.com/specimen/Source+Sans+Pro">Google Fonts page for the font</a> and clicked on “Select this font”. Then after clicking on “1 Family Selected” at the bottom, I downloaded the font by clicking on the download icon. This downloaded a zip drive into my Downloads folder. To install it as a font, I extracted the zip folder (by opening the zip drive). In a different Finder window, I opened my Applications folder and found Font Book. Then I dragged the zip folder over into Font Book.</p>
</div>
<div id="pc" class="section level3">
<h3>PC</h3>
<p>I modified the above instructions from these instructions on <a href="https://flourishonline.com.au/download-google-fonts-mac-pc/">Flourish Online</a>, which also has PC instructions. If anyone gives these a try and they don’t work, let me know and I’ll try to do some troubleshooting!</p>
</div>
</div>
<div id="font_add_google" class="section level2">
<h2><code>font_add_google</code></h2>
<p>Next, I used the <code>font_add_google</code> function from <code>sysfonts</code> to download the Source Sans Pro fonts:</p>
<pre class="r"><code>font_add_google(&quot;Source Sans Pro&quot;)</code></pre>
<p>If you want to install a different Google Font, you can run <code>font_families_google()</code> to see the list of family names of fonts currently available in Google Fonts:</p>
<pre class="r"><code>head(font_families_google())</code></pre>
<pre><code>## [1] &quot;ABeeZee&quot;       &quot;Abel&quot;          &quot;Abhaya Libre&quot;  &quot;Abril Fatface&quot;
## [5] &quot;Aclonica&quot;      &quot;Acme&quot;</code></pre>
</div>
<div id="example" class="section level2">
<h2>Example</h2>
<p>To continue with my Lego theme, I’ll demonstrate using this font on data from the <code>legocolors</code> package. We can make the following graph of approximate lego brick availability based on the year a brick was released (colored by the brick color!):</p>
<pre class="r"><code>library(legocolors)</code></pre>
<pre><code>## Warning: package &#39;legocolors&#39; was built under R version 3.5.2</code></pre>
<pre class="r"><code>library(ggplot2)</code></pre>
<pre><code>## Warning: package &#39;ggplot2&#39; was built under R version 3.5.2</code></pre>
<pre class="r"><code>ggplot(legocolors, aes(x = year_released, y = bl_bp, colour = hex)) +
  geom_point() +
  scale_color_manual(values = legocolors$hex) +
  theme_minimal() +
  theme(legend.position = &quot;none&quot;, text = element_text(family = &quot;Source Sans Pro&quot;), plot.title = element_text(face = &quot;bold&quot;)) +
  labs(x = &quot;Year released&quot;, y = &quot;Brick availability&quot;, title = &quot;Lego availability&quot;, subtitle = &quot;All of the text is now in Source Sans Pro!&quot;)</code></pre>
<pre><code>## Warning: Removed 3 rows containing missing values (geom_point).</code></pre>
<p><img src="/post/2019-10-11-google_fonts_files/figure-html/unnamed-chunk-4-1.png" width="672" /></p>
</div>
<div id="troubleshooting" class="section level2">
<h2>Troubleshooting</h2>
<p>This process is a little bit finicky, so there are certain errors that crop up frequently. For me, just restarting RStudio and/or my computer (or updating the computer, if necessary) has solved these. The most common errors are <code>No font could be found for family &quot;Source Sans Pro&quot;</code> and <code>&quot;Error in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, : polygon edge not found&quot;</code>, both of which relate to R not being able to find the font. If the font is indeed installed on your computer (you can check this in Font Book, for a Mac), then restarting things should help R find the font.</p>
</div>

    </div>
  
</div>

    </div>
  </section>

  
  
  

  

  

  

  

  

  
  

  
  
  

  
  
  
  
  

  
  

  <section id="2019-09-26-mode_api_instructions" class="home-section wg-blank   "  >
    <div class="container">
      


<div class="row">
  
    <div class="col-12 col-lg-4 section-heading">
      <h1>Accessing Mode data in R</h1>
      
    </div>
    <div class="col-12 col-lg-8">
      


<figure>
<img class="special-img-class" style="width:60%" src="/img/legos.jpg" />
<figcaption>
A lego fort that I constructed this summer with the help of a good friend (not particularly relevant to the post, except that I used a lego dataset for my example and legos are more photogenic than Mode databases).
</figcaption>
</figure>
<div id="inspiration-for-this-post" class="section level2">
<h2>Inspiration for this post</h2>
<p>In the past, I’ve been used to a pretty straightforward R workflow: download my data as a csv from somewhere, save it on my computer, and go to work. Now that I work in healthcare, downloading data and saving it on my computer is no longer such a good idea (since that would be illegal) so I’ve needed other ways to get the data I need into R. My company uses the data analysis platform <a href="https://mode.com/">Mode Analytics</a> to pull data from our database using SQL. I wanted a way to get data directly from Mode into R without needing to download the results of a report as a CSV, move them onto Google Drive, and then access them in R.</p>
</div>
<div id="get-api-access-token" class="section level2">
<h2>Get API access token</h2>
<p>The first step to accessing Mode via API is to <a href="https://mode.com/help/articles/user-accounts/#api-tokens">generate a token</a>. First, log in to Mode, click on your name in the top left corner, and select “My Account”. Then click on “API Tokens” on the left side. You can generate a token by specifying a name for the token and clicking “Create token”. The resulting credentials consist of your token (a.k.a. username or access key) and your password (a.k.a. access secret). This token is associated with your Mode user account, not an organization, so you’ll be able to access reports from any organization you have access to.</p>
</div>
<div id="export-report-csv-to-dataframe" class="section level2">
<h2>Export report CSV to dataframe</h2>
<p>The following instructions essentially follow the Python code instructions on the Mode website to <a href="https://mode.com/developer/api-cookbook/distribution/export-csv/">export a report to a csv</a>. This code doesn’t actually download the csv file, however, just imports it into R as a dataframe (so no datafiles on your computer!).</p>
<p>The following functions rely on the <code>httr</code> and <code>jsonlite</code> packages:</p>
<pre class="r"><code>library(httr)
library(jsonlite)</code></pre>
<p>You also need your username and password, preferably assigned to variables in R. For this example, my username is assigned to <code>username</code> and my password is assigned to <code>password</code>.</p>
<pre class="r"><code># Replace with your username and password
username &lt;- &quot;YOUR_USERNAME&quot;
password &lt;- &quot;YOUR_PASSWORD&quot;</code></pre>
<p>You also need to find the token of the report you’re interested in. When you go to your report in Mode, the token is everything after “reports/”. So, for example, if you’re interested in Legos (which you should be), you can get the most recent run of the <a href="https://app.mode.com/modeanalytics/reports/eb4fdefe37d8">Lego set summary</a> report. The URL is <a href="https://app.mode.com/modeanalytics/reports/eb4fdefe37d8" class="uri">https://app.mode.com/modeanalytics/reports/eb4fdefe37d8</a>, so the report token is “eb4fdefe37d8”.</p>
<pre class="r"><code># Replace with the report token of interest
report_token &lt;- &quot;eb4fdefe37d8&quot;</code></pre>
<p>Finally, you need the name of your organization on Mode. For this example, we’ll be looking at a community available Mode report, so the organization is “modeanalytics”. You can get this from the URL as well, it will be the part directly before “/reports”:</p>
<pre class="r"><code># Replace with the organization of interest
organization &lt;- &quot;modeanalytics&quot;</code></pre>
<div id="get-the-latest-run-token" class="section level3">
<h3>Get the latest run token</h3>
<p>The first step is to find the run token of the most recent time the Mode report was run. The following function gets that token, or throws an error if the latest run of the report was not successful:</p>
<pre class="r"><code>get_report_latest_run &lt;- function(report_token, organization, username, password){
  # Generate the URL for runs of the report you&#39;re interested in
  url = paste0(&quot;https://modeanalytics.com/api/&quot;, organization, &quot;/reports/&quot;, report_token, &quot;/runs&quot;)
  r = GET(url, authenticate(username, password, type = &quot;basic&quot;))
  result = fromJSON(content(r, &quot;text&quot;))
  # Find the most recent run token
  most_recent_report_run_token &lt;- result$`_embedded`$report_runs[1,]$token
  # Find the status of the most recent run
  status = result$`_embedded`$report_runs[1,]$state
  if(status == &quot;succeeded&quot;){ # if the most recent report ran,
    return(most_recent_report_run_token) # return that run token
  } else{
    stop(&quot;Report latest run did not succeed.&quot;) # if the most recent run failed, error
  }
}</code></pre>
<p>So, for our example report, the most recent run token was:</p>
<pre class="r"><code>get_report_latest_run(report_token, organization, username, password)</code></pre>
<pre><code>## No encoding supplied: defaulting to UTF-8.</code></pre>
<pre><code>## [1] &quot;079bc1243d11&quot;</code></pre>
</div>
<div id="extract-the-data-from-the-most-recent-run" class="section level3">
<h3>Extract the data from the most recent run</h3>
<p>Now, we can use a second function (which calls <code>get_report_latest_run</code>) to get the data from the most recent run:</p>
<pre class="r"><code>export_report_to_csv &lt;- function(report_token, organization, username, password) {
  # Get the latest run token
  latest_report_run_token &lt;- get_report_latest_run(report_token, organization, username, password)
  # Generate the URL for the latest run of the report as a csv
  url &lt;- paste0(&quot;https://modeanalytics.com/api/&quot;, organization, &quot;/reports/&quot;, report_token, &quot;/runs/&quot;, latest_report_run_token, &quot;/results/content.csv&quot;)
  r &lt;- GET(url, authenticate(username, password, type = &quot;basic&quot;))
  # Read the data as a csv
  r_csv &lt;- content(r, type = &quot;text/csv&quot;)
  return(r_csv)
}</code></pre>
</div>
</div>
<div id="example" class="section level2">
<h2>Example</h2>
<p>To get the data from the most recent run of the <a href="https://app.mode.com/modeanalytics/reports/eb4fdefe37d8">Lego set summary</a> report, you want to save a script file containing the above two functions (I call this script <code>mode_api_script.R</code>. Then, run the following:</p>
<pre class="r"><code>library(httr)
library(jsonlite)
source(&quot;mode_api_script.R&quot;)

# Set username, password, and report token
username &lt;- &quot;YOUR_USERNAME&quot;
password &lt;- &quot;YOUR_PASSWORD&quot;
report_token &lt;- &quot;eb4fdefe37d8&quot;</code></pre>
<pre class="r"><code>lego_data &lt;- export_report_to_csv(report_token, organization, username, password)</code></pre>
<pre><code>## No encoding supplied: defaulting to UTF-8.
## No encoding supplied: defaulting to UTF-8.</code></pre>
<pre><code>## Parsed with column specification:
## cols(
##   year = col_integer(),
##   white = col_double(),
##   red = col_double(),
##   blue = col_double(),
##   black = col_double(),
##   light_gray = col_double(),
##   total = col_double(),
##   main = col_character()
## )</code></pre>
<pre class="r"><code>knitr::kable(head(lego_data))</code></pre>
<table>
<thead>
<tr class="header">
<th align="right">year</th>
<th align="right">white</th>
<th align="right">red</th>
<th align="right">blue</th>
<th align="right">black</th>
<th align="right">light_gray</th>
<th align="right">total</th>
<th align="left">main</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="right">1950</td>
<td align="right">6</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">6</td>
<td align="left">White was the main color</td>
</tr>
<tr class="even">
<td align="right">1953</td>
<td align="right">12</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">12</td>
<td align="left">White was the main color</td>
</tr>
<tr class="odd">
<td align="right">1954</td>
<td align="right">6</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">6</td>
<td align="left">White was the main color</td>
</tr>
<tr class="even">
<td align="right">1955</td>
<td align="right">21</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">21</td>
<td align="left">White was the main color</td>
</tr>
<tr class="odd">
<td align="right">1956</td>
<td align="right">9</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">9</td>
<td align="left">White was the main color</td>
</tr>
<tr class="even">
<td align="right">1957</td>
<td align="right">45</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">45</td>
<td align="left">White was the main color</td>
</tr>
</tbody>
</table>
</div>

    </div>
  
</div>

    </div>
  </section>

  
  
  

  

  

  

  

  

  
  

  
  
  

  
  
  
  
  

  
  

  <section id="2019-09-18-git-workflow1" class="home-section wg-blank   "  >
    <div class="container">
      


<div class="row">
  
    <div class="col-12 col-lg-4 section-heading">
      <h1>Git Workflow #1: Existing project, GitHub last</h1>
      
    </div>
    <div class="col-12 col-lg-8">
      


<figure>
<img class="special-img-class" style="width:60%" src="/img/xkcd_git.png" />
<figcaption>
From <a href="https://xkcd.com/1597/">xkcd</a>.
</figcaption>
</figure>
<div id="inspiration-for-this-post" class="section level2">
<h2>Inspiration for this post</h2>
<p>Over the coming months, I plan on sharing a series of Git workflow tutorials. Getting a Git project set up can be a pretty simple process, but depending on the order in which you do things, set-up can get complicated. This first tutorial addresses one of those fairly complicated situations.</p>
<p>The schema for these tutorials is loosely based on the incredible Git + R bible by Jenny Bryan, <a href="https://happygitwithr.com/">Happy Git and GitHub for the useR</a>. The only difference is that I will be showing the workflow for setting up the project using the command line, rather than RStudio.</p>
</div>
<div id="existing-project-github-last" class="section level2">
<h2>Existing project, GitHub last</h2>
<p>This first tutorial addresses the following situation: you have a folder for your project on your computer already. You would like this folder to be associated with a GitHub repository, but you haven’t created this repository yet, or you’ve created the GitHub repository and haven’t connected it with your local folder [^1].</p>
<p>[^1] Throughout this tutorial, “local” refers to the copy of the repository on your computer. “Remote” refers to the copy on GitHub.</p>
<div id="make-a-git-repo" class="section level3">
<h3>Make a Git repo</h3>
<p>The first step in this process is to make your local project into a Git repository. Start by using the command-line to navigate to the folder for your project (using <code>cd</code>). Once you are in the proper folder, run:</p>
<pre class="bash"><code>git init</code></pre>
<p>This command either creates a new Git repository, or, in our case, converts an existing project into a Git repository. If you want to read more about <code>git init</code>, particularly the difference between <code>git init</code> and <code>git clone</code>, Atlassian has a useful <a href="https://www.atlassian.com/git/tutorials/setting-up-a-repository/git-init">article</a>.</p>
<p>Now, since you’ve made edits / added files into your project, you want to add and commit those changes using the following commands:</p>
<pre class="bash"><code>git add -A
git commit -m &quot;test commit&quot;</code></pre>
<p>You can make the commit message whatever you want. Note, <code>git add -A</code> adds all files in your repository. If you only want to add certain files, you can add them by name. For example, if you only want to add the file <code>test.txt</code>:</p>
<pre class="bash"><code>git add test.txt
git commit -m &quot;test commit&quot;</code></pre>
</div>
<div id="make-a-github-repo" class="section level3">
<h3>Make a GitHub repo</h3>
<p>Next, we need to create a GitHub repository for our project. Depending on your workflow, you may have already created this repository. If not, log in to <a href="https://github.com">GitHub</a>. Click the green “New repository” button. Ideally, you want the name of your new repository to match the name of the folder on your computer (but if not, it’s not a big deal). You don’t really want to initialize the repository with a README, but if you do, you can work around this later.</p>
<p>Once the repository has been set up, you can copy the HTTPS URL to your clipboard using the green “Clone or Download” button.</p>
</div>
<div id="connect-the-github-repo-to-your-project" class="section level3">
<h3>Connect the GitHub repo to your project</h3>
<p>Now that you have a GitHub repository and you have a Git repository on your computer, you want to connect the two. To do this, run the following in the shell, using the URL that you just copied from GitHub:</p>
<pre class="bash"><code>git remote add origin &lt;YOUR URL HERE&gt;</code></pre>
<p>Note, you don’t need the <code>&lt;</code> and <code>&gt;</code> on either side of the URL, just paste the URL as is.</p>
<p>You now need to tell your computer what the upstream branch should be, i.e. what branch on GitHub to pull updates from.</p>
<pre class="bash"><code>git branch --set-upstream-to=origin/master master</code></pre>
</div>
<div id="reconcile-the-remote-and-local-repositories" class="section level3">
<h3>Reconcile the remote and local repositories</h3>
<p>Now, assuming that you have edits both on your computer and on the GitHub repository, you need to be able to reconcile these changes so that the two are the same. This is assuming that you have files added in the folder on your computer and also something like a README or .gitignore on GitHub. To fix this problem, you first need to pull the changes from the GitHub repository onto your computer:</p>
<pre class="bash"><code>git pull --allow-unrelated-histories</code></pre>
<p>The <code>--allow-unrelated-histories</code> portion of the command allows you to merge these two repositories (the local and the remote) even though they don’t have a common ancestor (you created them independently and made changes to both independently).</p>
<p>Now, you can push the changes from your local repository up to GitHub:</p>
<pre class="bash"><code>git push</code></pre>
</div>
</div>

    </div>
  
</div>

    </div>
  </section>

  
  
  

  

  

  

  

  

  
  

  
  
  

  
  
  
  
  

  
  

  <section id="2019-08-21-goldwater-highlight" class="home-section wg-blank   "  >
    <div class="container">
      


<div class="row">
  
    <div class="col-12 col-lg-4 section-heading">
      <h1>Goldwater Highlight</h1>
      
    </div>
    <div class="col-12 col-lg-8">
      <figure>
  <img class="special-img-class" style="width:100%" src="/img/goldwater_banner.png" />
</figure>

<p>This is just a quick post to provide a <a href="https://twitter.com/GoldwaterSC/status/1163980565978259456" target="_blank">link</a> to the Goldwater Scholar Social Media Highlight that Nina Singh wrote about me. You can find the pdf to the article <a href="/goldwater_highlight.pdf">here</a>. Huge thanks to the Goldwater social media team for including me in this!</p>

    </div>
  
</div>

    </div>
  </section>

  
  
  

  

  

  

  

  

  
  

  
  
  

  
  
  
  
  

  
  

  <section id="2019-06-07-usb" class="home-section wg-blank   "  >
    <div class="container">
      


<div class="row">
  
    <div class="col-12 col-lg-4 section-heading">
      <h1>Copying R libraries to a USB</h1>
      
    </div>
    <div class="col-12 col-lg-8">
      


<figure>
<img class="special-img-class" style="width:60%" src="/img/usb_meme.jpg" />
<figcaption>
From <a href="https://www.facebook.com/groups/WildGreenMemesForEcologicalFiends/">Wild Green Memes for Ecological Fiends</a>.
</figcaption>
</figure>
<div id="inspiration-for-this-post" class="section level2">
<h2>Inspiration for this post</h2>
<p>A few people in my lab are headed to a workshop in Colombia next week, and due to the possibility of intermittent wi-fi, they need to load all of the workshop materials onto USB drives. They need R with certain packages on the USBs in versions compatible with both Macs and PCs, so I offered to help out by loading my Mac version of R and my packages.</p>
</div>
<div id="loading-r-onto-a-usb" class="section level2">
<h2>Loading R onto a USB</h2>
<p>The first step was to copy my installation of R onto the USB. To do this, I just copied R from my Applications folder onto the folder. We are not entirely sure if this will work yet to transfer R to someone else’s computer (if someone knows the answer, let me know!), but we are going to test it out shortly.</p>
</div>
<div id="loading-r-packages-onto-a-usb" class="section level2">
<h2>Loading R packages onto a USB</h2>
<p>The next step was to transfer all of my R packages onto the USB. To do this, I first needed to figure out where the packages were saved on my computer. You can run the following in R to find where your packages are saved:</p>
<pre class="r"><code>.libPaths()
## [1] &quot;/Library/Frameworks/R.framework/Versions/3.5/Resources/library&quot;</code></pre>
<p>Next, I needed to figure out how to navigate to the USB in Terminal. According to this <a href="https://apple.stackexchange.com/questions/60231/using-terminal-how-can-i-find-which-directory-is-my-usb-drive-mounted-in">StackOverflow answer</a>, all drives are mounted in <code>/Volumes</code>, so I ran the following in Terminal:</p>
<pre class="bash"><code>cd /Volumes
ls</code></pre>
<p>From there, I was able to see <code>USB DISK</code> as one of the folders in <code>/Volumes</code>. In order to copy all of the packages from my computer onto the USB, I used <code>cp</code> to move everything from the folder found in <code>.libPaths()</code> onto the USB:</p>
<pre class="bash"><code># replace the first file path with what you get from .libPaths()
cp -R /Library/Frameworks/R.framework/Versions/3.5/Resources/library/ /Volumes/USB\ DISK/R_packages/</code></pre>
</div>

    </div>
  
</div>

    </div>
  </section>

  
  
  

  

  

  

  

  

  
  

  
  
  

  
  
  
  
  

  
  

  <section id="2019-05-23-highres-figures" class="home-section wg-blank   "  >
    <div class="container">
      


<div class="row">
  
    <div class="col-12 col-lg-4 section-heading">
      <h1>Setting resolution and aspect ratios in R</h1>
      
    </div>
    <div class="col-12 col-lg-8">
      


<figure>
<img class="special-img-class" style="width:60%" src="/img/resolution_comic.jpg" />
<figcaption>
Bizarro.
</figcaption>
</figure>
<div id="inspiration-for-this-post" class="section level2">
<h2>Inspiration for this post</h2>
<p>I recently needed to make some figures to present and noticed that the ones I was importing from saved R plots were showing up fuzzy on the presentation. Shortly afterwards, I had interns trying to generate figures with the same aspect ratio from different computers. I did a little digging and found this method for saving figures with specified resolutions and aspect ratios.</p>
</div>
<div id="solution" class="section level2">
<h2>Solution</h2>
<p>In order to save your figure with a certain size and resolution, you just need to include your code for plotting between the two lines below:</p>
<pre class="r"><code># To save a .png file:
png(&quot;your_image.png&quot;, units = &quot;in&quot;, width = 5, height = 4, res = 300)
# your plotting code here
dev.off()</code></pre>
<p>You can also save .bmp, .jpeg, and .tiff files in the same way with the <code>bmp</code>, <code>jpeg</code> and <code>tiff</code> functions.</p>
</div>
<div id="example" class="section level2">
<h2>Example</h2>
<p>To see this in practice, we can plot some data from the iris dataset the “regular” way:</p>
<pre class="r"><code>pairs(iris[,1:4], col = iris$Species)</code></pre>
<p><img src="/post/2019-05-23-highres-figures_files/figure-html/unnamed-chunk-2-1.png" width="672" /></p>
<p>Things look pretty nice in the plotting window, but when you save that figure from RStudio, this is the result:</p>
<figure>
<img class="special-img-class" style="width:80%" src="/img/lowres.png" />
</figure>
<p>To get rid of that graininess, we can instead save the figure in the following way:</p>
<pre class="r"><code>png(&quot;highres.png&quot;, units = &quot;in&quot;, width = 5, height = 4, res = 300)
pairs(iris[,1:4], col = iris$Species)
dev.off()</code></pre>
<p>The resulting figure is much clearer and ready to be inserted in presentations or publications!</p>
<figure>
<img class="special-img-class" style="width:80%" src="/img/highres.png" />
</figure>
<p>You can also change the aspect ratio by modifying the <code>height</code> and <code>width</code> arguments. You can leave the units in inches or set it to pixels (<code>&quot;px&quot;</code>), centimeters (<code>&quot;cm&quot;</code>), or millimeters (<code>&quot;mm&quot;</code>).</p>
</div>

    </div>
  
</div>

    </div>
  </section>

  
  
  

  

  

  

  

  

  
  

  
  
  

  
  
  
  
  

  
  

  <section id="2019-04-12-sdm-threshold" class="home-section wg-blank   "  >
    <div class="container">
      


<div class="row">
  
    <div class="col-12 col-lg-4 section-heading">
      <h1>Thresholding species distribution models</h1>
      
    </div>
    <div class="col-12 col-lg-8">
      


<figure>
<img class="special-img-class" style="width:100%" src="/img/thresholds.png" />
<figcaption>
From left to right: unmodified species distribution model, minimum training presence threshold, and 10th percentile threshold.
</figcaption>
</figure>
<div id="inspiration-for-this-post" class="section level2">
<h2>Inspiration for this post</h2>
<p>Conservation is often the main motivation behind studying where a species lives – having a model of a species’ range can help scientists assess whether it is at risk of extinction, designate protected regions to preserve its habitat, and study potential impacts of human activity. When we create species distribution models using common methods like Maxent, the result is a map of predicted habitat suitability or probability of species presence, such as the one below. In conservation management, however, it is often more useful to present range models in the form of species presence/absence. We can convert continuous predictions of habitat suitability into binary predictions of whether a species lives in a certain region or not using thresholds: i.e. designating all regions above a certain suitability level as within the species range and all areas below that suitability level as outside of it.</p>
<figure>
<img class="special-img-class" style="width:60%" src="/img/threshold_ex.png" />
<figcaption>
Left: species distribution model with continuous habitat suitability values. Right: binary presence/absence model used by applying a threshold. (Figure from <a href="https://rspatial.org/sdm/6_sdm_methods.html#mahalanobis-distance">Spatial Data Science with R</a>)
</figcaption>
</figure>
<p>I recently needed to threshold some species distribution models to convert them into these binary maps and had difficulty finding a built-in way to do this in R. The <code>dismo</code> package for species distribution modeling has a function <code>threshold</code> to find what value to use as the “cut-off”, but I needed a function to apply a given cut-off value to model and output a raster with binary values for presence and absence.</p>
</div>
<div id="thresholding-function" class="section level2">
<h2>Thresholding function</h2>
<p>I wrote an R function to take a species distribution model and threshold it by a given threshold - either minimum training presence (MTP) or 10th percentile training present (P10).</p>
<div id="minimum-training-presence" class="section level3">
<h3>Minimum training presence</h3>
<p>This threshold finds the lowest predicted suitability value for an occurrence point. Essentially, it assumes that the least suitable habitat at which the species is known to occur is the minimum suitability value for the species. The MTP threshold ensures that all occurrence points fall within the area of the binary model.</p>
</div>
<div id="th-percentile-training-presence" class="section level3">
<h3>10th percentile training presence</h3>
<p>The P10, on the other hand, is a threshold which omits all regions with habitat suitability lower than the suitability values for the lowest 10% of occurrence records. It assumes that the 10% of occurrence records in the least suitable habitat aren’t occurring in regions that are representative of the species overall habitat, and thus should be omitted. This threshold omits a greater region than the MTP.</p>
</div>
<div id="the-function" class="section level3">
<h3>The function</h3>
<p>The following is the function I wrote to apply these two thresholds to an SDM. The function’s arguments are the SDM, the occurrence points of the species in the form of longitude - latitude pairs, the threshold type, and whether the user would like the output to be a binary prediction (0s for predicted absence and 1s for predicted presence), or a thresholded continuous SDM (regions with suitability below the threshold set to 0).</p>
<pre class="r"><code>library(raster)
## Warning: package &#39;raster&#39; was built under R version 3.5.2
## Loading required package: sp

sdm_threshold &lt;- function(sdm, occs, type = &quot;mtp&quot;, binary = FALSE){
  occPredVals &lt;- raster::extract(sdm, occs)
  if(type == &quot;mtp&quot;){
    thresh &lt;- min(na.omit(occPredVals))
  } else if(type == &quot;p10&quot;){
    if(length(occPredVals) &lt; 10){
      p10 &lt;- floor(length(occPredVals) * 0.9)
    } else {
      p10 &lt;- ceiling(length(occPredVals) * 0.9)
    }
    thresh &lt;- rev(sort(occPredVals))[p10]
  }
  sdm_thresh &lt;- sdm
  sdm_thresh[sdm_thresh &lt; thresh] &lt;- NA
  if(binary){
    sdm_thresh[sdm_thresh &gt;= thresh] &lt;- 1
  }
  return(sdm_thresh)
}</code></pre>
<p>The first step of the function is to extract the SDM predictions at all occurrence points.</p>
<pre class="r"><code>occPredVals &lt;- raster::extract(sdm, occs)</code></pre>
<p>Next, the function calculates a threshold value <code>thresh</code> for either the MTP or P10 threshold. Finally, it sets all cells in the SDM raster with values lower than the threshold equal to 0. If the user wants a binary map, the function sets all cells above the threshold equal to 1:</p>
<pre class="r"><code>sdm_thresh &lt;- sdm
sdm_thresh[sdm_thresh &lt; thresh] &lt;- NA
if(binary){
  sdm_thresh[sdm_thresh &gt;= thresh] &lt;- 1
}</code></pre>
</div>
</div>
<div id="example" class="section level2">
<h2>Example</h2>
<p>Now we can apply the function to an actual SDM I generated for a species of three-toed sloth (<em>Bradypus variegatus</em>).</p>
<pre class="r"><code># load in the SDM and occurrence points
sloth_sdm &lt;- raster(&quot;/Users/hellenfellows/Desktop/website-hugo/static/SDMs/variegatus_sdm.tif&quot;)
sloth_occs &lt;- read.csv(&quot;~/Desktop/website-hugo/static/SDMs/variegatus_occ.csv&quot;)

plot(sloth_sdm)
points(sloth_occs[,2:3], pch = 19, cex = 0.5)</code></pre>
<p><img src="/post/2019-04-12-sdm-threshold_files/figure-html/unnamed-chunk-4-1.png" width="672" /></p>
<p>We can apply both MTP and P10 thresholds to the SDM based on the location of the occurrence points:</p>
<pre class="r"><code>sloth_mtp &lt;- sdm_threshold(sloth_sdm, sloth_occs[,2:3], &quot;mtp&quot;)
plot(sloth_mtp)</code></pre>
<p><img src="/post/2019-04-12-sdm-threshold_files/figure-html/unnamed-chunk-5-1.png" width="672" /></p>
<pre class="r"><code>
sloth_p10 &lt;- sdm_threshold(sloth_sdm, sloth_occs[,2:3], &quot;p10&quot;)
plot(sloth_p10)</code></pre>
<p><img src="/post/2019-04-12-sdm-threshold_files/figure-html/unnamed-chunk-5-2.png" width="672" /></p>
<p>We could also make either of these thresholded SDMs into a binary prediction in the following way:</p>
<pre class="r"><code>sloth_mtp_bin &lt;- sdm_threshold(sloth_sdm, sloth_occs[,2:3], &quot;mtp&quot;, binary = TRUE)
plot(sloth_mtp_bin)</code></pre>
<p><img src="/post/2019-04-12-sdm-threshold_files/figure-html/unnamed-chunk-6-1.png" width="672" /></p>
</div>
<div id="generalization" class="section level2">
<h2>Generalization</h2>
<p>My primary motivation to write this function was to use it on SDMs, but the function could easily be generalized to threshold any raster by a given value:</p>
<pre class="r"><code>raster_threshold &lt;- function(input_raster, points = NULL, type = NULL, threshold = NULL, binary = FALSE) {
  if (!is.null(points)) {
    pointVals &lt;- raster::extract(input_raster, points)
    if (type == &quot;mtp&quot;) {
      threshold &lt;- min(na.omit(pointVals))
    } else if (type == &quot;p10&quot;) {
      if (length(pointVals) &lt; 10) {
        p10 &lt;- floor(length(pointVals) * 0.9)
      } else {
        p10 &lt;- ceiling(length(pointVals) * 0.9)
      }
      threshold &lt;- rev(sort(pointVals))[p10]
    }
  }
  raster_thresh &lt;- input_raster
  raster_thresh[raster_thresh &lt; threshold] &lt;- NA
  if (binary) {
    raster_thresh[raster_thresh &gt;= threshold] &lt;- 1
  }
  return(raster_thresh)
}</code></pre>
<p>I expanded the function to allow the user to input points within the raster to calculate MTP and P10 thresholds if desired, but also to enable a user-specified threshold.</p>
<pre class="r"><code># create arbitrary raster
raster1 &lt;- raster(nrow=10, ncol=10)
raster1[1:25]&lt;- 1:25
raster1[26:50] &lt;- rev(1:25)
raster1[51:75] &lt;- 1:25
raster1[76:100] &lt;- rev(1:25)

# create a set of 20 arbitrary points within the raster
xy &lt;- data.frame(x = runif(20, min = -150, max = 150), y = runif(20, min = -70, max = 70))

plot(raster1)
points(xy)</code></pre>
<p><img src="/post/2019-04-12-sdm-threshold_files/figure-html/unnamed-chunk-8-1.png" width="672" /></p>
<p>Now we can apply the function to see the MTP and P10 thresholded rasters:</p>
<pre class="r"><code>mtp_raster &lt;- raster_threshold(input_raster = raster1, points = xy, type = &quot;mtp&quot;, binary = TRUE)
plot(mtp_raster)</code></pre>
<p><img src="/post/2019-04-12-sdm-threshold_files/figure-html/unnamed-chunk-9-1.png" width="672" /></p>
<pre class="r"><code>
p10_raster &lt;- raster_threshold(input_raster = raster1, points = xy, type = &quot;p10&quot;, binary = TRUE)
plot(p10_raster)</code></pre>
<p><img src="/post/2019-04-12-sdm-threshold_files/figure-html/unnamed-chunk-9-2.png" width="672" /></p>
<p>We can also use a user-inputted threshold to remove all parts of the raster with values lower than 20:</p>
<pre class="r"><code>user_raster &lt;- raster_threshold(input_raster = raster1, threshold = 20)
plot(user_raster)</code></pre>
<p><img src="/post/2019-04-12-sdm-threshold_files/figure-html/unnamed-chunk-10-1.png" width="672" /></p>
</div>

    </div>
  
</div>

    </div>
  </section>

  
  
  

  

  

  

  

  

  
  

  
  
  

  
  
  
  
  

  
  

  <section id="2019-03-18-alpha-hull" class="home-section wg-blank   "  >
    <div class="container">
      


<div class="row">
  
    <div class="col-12 col-lg-4 section-heading">
      <h1>Converting alpha hulls to spatial objects</h1>
      
    </div>
    <div class="col-12 col-lg-8">
      


<figure>
<img class="special-img-class" style="width:60%" src="/img/alpha_bg_regions.png" />
<figcaption>
Alpha hulls (α = 20) around occurrence points for two species of sloths.
</figcaption>
</figure>
<div id="inspiration-for-this-post" class="section level2">
<h2>Inspiration for this post</h2>
<p>In species distribution modeling, one of the key steps requires the researcher to select a “background region” for the species, i.e. a region over which a machine learning model will compare the environment of the “background points” with the environment at points where the species is known to occur. The key to selecting this region is to pick an area where the species could occur but hasn’t necessarily been observed – for example, you don’t want to include an area separated from the rest of the range by a big mountain range that you don’t believe the organism could cross, but you do want to include a range of potential environments. There are many methods to delineate this region, from drawing a box around the occurrence points of the species to creating a buffered region around each occurrence point (think a collection of lots of circles around each point). One of my research mentors suggested that I try a new method using a shape known as an α-hull.</p>
<p>I’ll describe α-hulls in more detail below, but you can get the gist from the map above, where I show occurrence points for two species of sloths surrounded by an α-hull for each species. When I tried to create this region in R, however, I ran into a roadblock: the α-hull objects were a specific kind of R object that didn’t play nicely with spatial data in R, particularly objects from the <code>sp</code> package. In particular, I needed a way to convert α objects into SpatialPolygons. This post describes a series of functions I wrote to carry out this process.</p>
</div>
<div id="the-alphahull-package" class="section level2">
<h2>The <code>alphahull</code> package</h2>
<p>The <code>alphahull</code> R package (Pateiro-Lopez et al. 2016) draws shapes (like the ones above) around sets of points based on a given parameter, α. The package creates two kinds of shapes I was interested in: α-shapes and α-convex hulls. The functions in this post convert these shapes into objects compatible with the <code>sp</code> package, which can then be used for spatial analyses, including creating background regions for species distribution modeling.</p>
<pre class="r"><code># load packages
library(alphahull)
library(sp)</code></pre>
<div id="alpha-shapes" class="section level3">
<h3>Alpha shapes</h3>
<p>Alpha shapes consist of a collection of lines drawn around a group of points. Probably the most familiar example of an α-shape is a convex hull, the smallest convex shape that can be drawn around a group of objects. For example, the following code draws a convex hull around some data from the iris dataset:</p>
<pre class="r"><code>data(iris)
iris_sepals &lt;- iris[,1:2]
# remove duplicate datapoints
iris_sepals &lt;- iris_sepals[!duplicated(paste(iris_sepals$Sepal.Length, iris_sepals$Sepal.Width)), ]

# find points that lie on the convex hull
convexhull &lt;- chull(iris_sepals)
# plot the data points
plot(iris_sepals, pch = 19, col = &quot;darkseagreen&quot;)
hull_pts &lt;- c(convexhull, convexhull[1])
# plot the convex hull
lines(iris_sepals[hull_pts, ], col = &quot;magenta&quot;)</code></pre>
<p><img src="/post/2019-03-18-alpha-hull_files/figure-html/unnamed-chunk-2-1.png" width="672" /></p>
<p>This convex hull (drawn in magenta) is an example of an α-shape: all convex hulls are α-shapes, but not all α-shapes are convex hulls. An α-shape doesn’t have to be convex – the lines making up the border of the shape can create concave edges relative to the points in the dataset. For example:</p>
<pre class="r"><code># create a three-paneled figure
par(mfrow = c(1,3))

# create three different alpha shapes
alphashape_0.5 &lt;- ashape(iris_sepals, alpha = 0.5)
alphashape_1 &lt;- ashape(iris_sepals, alpha = 1)
alphashape_2 &lt;- ashape(iris_sepals, alpha = 2)

# plot alpha = 0.5
plot(iris_sepals, pch = 19, col = &quot;darkseagreen&quot;)
plot(alphashape_0.5, col = &quot;magenta&quot;, add = TRUE)
# plot alpha = 1
plot(iris_sepals, pch = 19, col = &quot;darkseagreen&quot;)
plot(alphashape_1, col = &quot;magenta&quot;, add = TRUE)
# plot alpha = 2
plot(iris_sepals, pch = 19, col = &quot;darkseagreen&quot;)
plot(alphashape_2, col = &quot;magenta&quot;, add = TRUE)</code></pre>
<p><img src="/post/2019-03-18-alpha-hull_files/figure-html/unnamed-chunk-3-1.png" width="672" /></p>
<pre class="r"><code># reset plotting parameters
par(mfrow = c(1,1))</code></pre>
<p>Alpha shapes are created using the <code>ashape</code> function from the <code>alphahull</code> package. As you can see, increasing the α value makes the shape closer and closer to the convex hull, while low values of α make the shape more concave.</p>
<div id="alpha-shapes-to-polygons" class="section level4">
<h4>Alpha shapes to polygons</h4>
<p>In order to use α-shapes with spatial data in R, I wanted to convert these shapes to polygons. To accomplish this, I modified some of the code from an <a href="https://rpubs.com/geospacedman/alphasimple">RPubs by Barry Rowlingson</a> to create the following function:</p>
<pre class="r"><code>library(igraph)
## 
## Attaching package: &#39;igraph&#39;
## The following objects are masked from &#39;package:stats&#39;:
## 
##     decompose, spectrum
## The following object is masked from &#39;package:base&#39;:
## 
##     union

ashape2poly &lt;- function(ashape){
  # Convert node numbers into characters
  ashape$edges[,1] &lt;- as.character(ashape$edges[,1])
  ashape_graph &lt;- graph_from_edgelist(ashape$edges[,1:2], directed = FALSE)
  if (!is.connected(ashape_graph)) {
    stop(&quot;Graph not connected&quot;)
  }
  if (any(degree(ashape_graph) != 2)) {
    stop(&quot;Graph not circular&quot;)
  }
  if (clusters(ashape_graph)$no &gt; 1) {
    stop(&quot;Graph composed of more than one circle&quot;)
  }
  # Delete one edge to create a chain
  cut_graph &lt;- ashape_graph - E(ashape_graph)[1]
  # Find chain end points
  ends = names(which(degree(cut_graph) == 1))
  path = get.shortest.paths(cut_graph, ends[1], ends[2])$vpath[[1]]
  # this is an index into the points
  pathX = as.numeric(V(ashape_graph)[path]$name)
  # join the ends
  pathX = c(pathX, pathX[1])
  return(pathX)
}</code></pre>
<p>For the reasoning behind the function, check out the RPubs I referred to for guidance. For a sanity check, we can compare the resulting shape to the original α-shape we were trying to replicate:</p>
<pre class="r"><code>alphapoly_1 &lt;- ashape2poly(alphashape_1)
plot(iris_sepals, pch = 19, col = &quot;darkseagreen&quot;)
# show the original alpha shape
plot(alphashape_1, lwd = 5, col = &quot;gray&quot;, add = TRUE)
# plot the new polygon
lines(iris_sepals[alphapoly_1, ], col = &quot;magenta&quot;)</code></pre>
<p><img src="/post/2019-03-18-alpha-hull_files/figure-html/unnamed-chunk-5-1.png" width="672" /></p>
</div>
</div>
<div id="alpha-hulls" class="section level3">
<h3>Alpha hulls</h3>
<p>Alpha hulls add another layer of complexity to this process because they can include curved lines (arcs) as edges of a shape. For example:</p>
<pre class="r"><code>alphahull_1 &lt;- ahull(iris_sepals, alpha = 1)

plot(iris_sepals, pch = 19, col = &quot;darkseagreen&quot;)
plot(alphahull_1, col = &quot;magenta&quot;, add = TRUE)</code></pre>
<p><img src="/post/2019-03-18-alpha-hull_files/figure-html/unnamed-chunk-6-1.png" width="672" /></p>
<div id="arcs-to-lines" class="section level4">
<h4>Arcs to lines</h4>
<p>To deal with this curvature, I wrote the following function to convert the arcs between points in the hull to a series of very short line segments in order to approximate the curve.</p>
<pre class="r"><code># function to convert an arc into line segments
# given the center of the arc, the radius, the vector, and the angle (radians)
arc2line &lt;- function(center, r, vector, theta, npoints = 100) {
  # Get the angles at the extremes of the arcs
  angles &lt;- anglesArc(vector, theta)
  # Generate sequence of angles along the arc to determine the points
  seqang &lt;- seq(angles[1], angles[2], length = npoints)
  # Generate x coordinates for points along the arc
  x &lt;- center[1] + r * cos(seqang)
  # Generate y coordinates for points along the arc
  y &lt;- center[2] + r * sin(seqang)
  coords.xy &lt;- cbind(x,y)
  line &lt;- Line(coords = coords.xy)
  return(line)
}</code></pre>
</div>
<div id="hulls-to-lines" class="section level4">
<h4>Hulls to lines</h4>
<p>Using the previous function, I wrote another function to take an α-hull and convert it into a set of SpatialLines objects. The function uses the <code>arc2line</code> function from above to convert each arc in the α-hull into a series of lines, before adding each of these sets of lines together.</p>
<pre class="r"><code>ahull2lines &lt;- function(hull){
  arclist &lt;- hull$arcs
  lines &lt;- list()
  for (i in 1:nrow(arclist)) {
    # Extract the attributes of arc i
    center_i &lt;- arclist[i, 1:2]
    radius_i &lt;- arclist[i, 3]
    vector_i &lt;- arclist[i, 4:5]
    theta_i &lt;- arclist[i, 6]
    # Convert arc i into a Line object
    line_i &lt;- arc2line(center = center_i, r = radius_i, vector = vector_i, theta = theta_i)
    list_length &lt;- length(lines)
    if(list_length &gt; 0){
      # If a line has already been added to the list of lines
      # Define last_line_coords as the coordinates of the last line added to the list before the ith line
      last_line_coords &lt;- lines[[list_length]]@coords
    }
    if(i == 1){
      # Add the first line to the list of lines
      lines[[i]] &lt;- line_i
    } else if(all.equal(line_i@coords[1,], last_line_coords[nrow(last_line_coords),])){
      # If the first coordinate in the ith line is equal to the last coordinate in the previous line
      # then those lines should be connected
      # Row bind the coordinates for the ith line to the coordinates of the previous line in the list
      lines[[list_length]]@coords &lt;- rbind(last_line_coords, line_i@coords[2:nrow(line_i@coords),])
    } else {
      # If the first coordinate in the ith line does not match the last coordinate in the previous line
      # then the ith line represents a new line
      # Add the ith line to the list as a new element
      lines[[length(lines) + 1]] &lt;- line_i
    }
  }
  # Convert the list of lines to a Line object
  lines &lt;- Lines(lines, ID = &#39;l&#39;)
  # Convert the Line object to a SpatialLines object
  sp_lines &lt;- SpatialLines(list(lines))
  return(sp_lines)
}</code></pre>
<p>The results look like this:</p>
<pre class="r"><code>lines_1 &lt;- ahull2lines(alphahull_1)
# the result is a SpatialLines object
class(lines_1)
## [1] &quot;SpatialLines&quot;
## attr(,&quot;package&quot;)
## [1] &quot;sp&quot;

plot(iris_sepals, pch = 19, col = &quot;darkseagreen&quot;)
# show the original alpha shape
plot(alphahull_1, lwd = 5, col = &quot;gray&quot;, add = TRUE)
# plot the new polygon
plot(lines_1, col = &quot;magenta&quot;, add = TRUE)</code></pre>
<p><img src="/post/2019-03-18-alpha-hull_files/figure-html/unnamed-chunk-9-1.png" width="672" /></p>
<p>The resulting SpatialLines object is an almost spot-on approximation of the original α-hull (shown in gray).</p>
</div>
<div id="spatiallines-to-spatialpolygon" class="section level4">
<h4>SpatialLines to SpatialPolygon</h4>
<p>Now, I needed a way to convert the SpatialLines object into a SpatialPolygon that would cover the same shape as the original α-hull. To accomplish this, I wrote a function that takes a SpatialLines object, checks which lines are part of polygons (i.e. form closed shapes), and converts those polygons to a SpatialPolygon.</p>
<pre class="r"><code>spLines2poly &lt;- function(sp_lines){
  # Extract the lines slot
  lines_slot &lt;- sp_lines@lines[[1]]
  # Create a list of booleans indicating whether a given Line represents a polygon
  poly_bool &lt;- sapply(lines_slot@Lines, function(x){
    coords &lt;- lines_slot@Lines[[1]]@coords
    # Check if the first coordinate in the line is the same as the last
    all.equal(coords[1,], coords[nrow(coords),])
  })
  # Pull out the lines that form polygons
  poly_lines &lt;- sp_lines[poly_bool]
  poly_lines_slot &lt;- poly_lines@lines
  # Create SpatialPolygons
  sp_polys &lt;- SpatialPolygons(list(Polygons(lapply(poly_lines_slot, function(x) {
    Polygon(slot(slot(x, &quot;Lines&quot;)[[1]], &quot;coords&quot;))
  }), ID = &quot;1&quot;)))
  return(sp_polys)
}</code></pre>
<p>We can apply this function to the <code>lines_1</code> object we made from the original <code>alphahull_1</code>:</p>
<pre class="r"><code>SpPoly_1 &lt;- spLines2poly(lines_1)
class(SpPoly_1)
## [1] &quot;SpatialPolygons&quot;
## attr(,&quot;package&quot;)
## [1] &quot;sp&quot;

plot(iris_sepals, pch = 19, col = &quot;darkseagreen&quot;)
# show the original alpha shape
plot(alphahull_1, lwd = 5, col = &quot;gray&quot;, add = TRUE)
# plot the new polygon
plot(SpPoly_1, border = &quot;magenta&quot;, add = TRUE)</code></pre>
<p><img src="/post/2019-03-18-alpha-hull_files/figure-html/unnamed-chunk-11-1.png" width="672" /></p>
</div>
<div id="alpha-hulls-to-spatialpolygons" class="section level4">
<h4>Alpha hulls to SpatialPolygons</h4>
<p>Finally, we can string all of those functions together to create a single function that will convert an α-hull directly into a SpatialPolygon:</p>
<pre class="r"><code>ahull2poly &lt;- function(hull){
  # Convert the alpha hull to SpatialLines
  hull2SpatialLines &lt;- ahull2lines(hull)
  # Convert SpatialLines to SpatialPolygon
  SpatialLines2SpatialPolygon &lt;- spLines2poly(hull2SpatialLines)
  return(SpatialLines2SpatialPolygon)
}</code></pre>
<p>As a final sanity check, we can see that the resulting shape is the same as the original shape produced by the <code>alphahull</code> package:</p>
<pre class="r"><code>hullpoly_1 &lt;- ahull2poly(alphahull_1)
class(hullpoly_1)
## [1] &quot;SpatialPolygons&quot;
## attr(,&quot;package&quot;)
## [1] &quot;sp&quot;

plot(iris_sepals, pch = 19, col = &quot;darkseagreen&quot;)
# show the original alpha shape
plot(alphahull_1, lwd = 5, col = &quot;gray&quot;, add = TRUE)
# plot the new polygon
plot(hullpoly_1, border = &quot;magenta&quot;, add = TRUE)</code></pre>
<p><img src="/post/2019-03-18-alpha-hull_files/figure-html/unnamed-chunk-13-1.png" width="672" /></p>
</div>
</div>
</div>
<div id="github" class="section level2">
<h2>GitHub</h2>
<p>The code for these functions is on my GitHub at <a href="https://github.com/babichmorrowc/hull2spatial" class="uri">https://github.com/babichmorrowc/hull2spatial</a>. If you have any thoughts or suggestions, please comment on this post or submit a pull request on GitHub. I hope to be formulating these functions into a package in the near future, so stay tuned!</p>
</div>
<div id="citations" class="section level2">
<h2>Citations</h2>
<p>Beatriz Pateiro-Lopez and Alberto Rodriguez-Casal. (2016). alphahull: Generalization of the Convex Hull of a Sample of Points in the Plane. R package version 2.1. <a href="https://CRAN.R-project.org/package=alphahull" class="uri">https://CRAN.R-project.org/package=alphahull</a></p>
</div>

    </div>
  
</div>

    </div>
  </section>

  
  
  

  

  

  

  

  

  
  

  
  
  

  
  
  
  
  

  
  

  <section id="2019-02-20-meteoritecluster1" class="home-section wg-blank   "  >
    <div class="container">
      


<div class="row">
  
    <div class="col-12 col-lg-4 section-heading">
      <h1>Hacking Meteorites Part 1: Calculating percent weights</h1>
      
    </div>
    <div class="col-12 col-lg-8">
      


<figure>
<img class="special-img-class" style="width:100%" src="/img/obj1_8bt_Fe.png" />
<figcaption>
Intensity of iron in a meteorite.
</figcaption>
</figure>
<p>Earlier this month I participated in the American Museum of Natural History’s annual hackathon (see my last <a href="https://babichmorrowc.github.io/post/2019-02-14-hackathon/">post</a> for details about the event overall). During the hackathon, I worked with Katy Abbott, another Helen Fellow at the museum with me, to use a machine learning algorithm called DBSCAN to tackle our challenge. This post explains the process we used to complete the first step of the challenge: estimating the percent weights of elements in a meteorite.</p>
<div id="challenge-accepted" class="section level2">
<h2>Challenge accepted</h2>
<p>Our team, consisting of Peter Kang, Jackson Lee, Jeremy Neiman, John Underwood, Katy Abbott, Meret Götschel, and myself, chose to work on the <a href="https://github.com/amnh/HackTheSolarSystem/wiki/Meteorite-Mineral-Mapping">Meteorite Mineral Mapping challenge</a>. For this challenge, our museum stakeholders, Marina Gemma and Sam Alpert, wanted a way to identify the mineral composition of meteorites.</p>
</div>
<div id="pixels-to-percents" class="section level2">
<h2>Pixels to percents</h2>
<p>The scientists scan meteorites with an electron microprobe, a device that provides the intensity of x-rays emitted from certain elements. The electron microprobe results in images with grayscale intensities corresponding to these x-ray intensities on a pixel-by-pixel basis.</p>
<figure>
<img class="special-img-class" style="width:100%" src="/img/meteorite_pixel_intensity.png" />
<figcaption>
Images of a meteorite produced by the electron microprobe showing pixel intensities for each of 10 elements - brighter grayscale values indicate greater amounts of that element in a pixel.
</figcaption>
</figure>
<p>Our first step was to convert these pixel intensities to the percent weight of each element in the mineral at that pixel. To do that, we referred to a set of standard images taken of minerals with a known proportion of each element. For example, the image below shows the pixel intensities of iron in 8 minerals: you can see that the pixels are brightest in Fe, or pure iron, and lower in iron oxide (<span class="math inline">\(\text{Fe}_3\text{O}_4\)</span>) and troilite, or iron sulfide (FeS).</p>
<figure>
<img class="special-img-class" style="width:100%" src="/img/iron_standard.png" />
<figcaption>
Standard scan of the intensity of iron in each of 8 known minerals.
</figcaption>
</figure>
<p>We can relate the intensity of the iron pixels in the Fe mineral, for example, to the percent weight in that mineral (100%, since iron is the only element in Fe).</p>
<div id="import-data" class="section level3">
<h3>Import data</h3>
<p>We started by creating a .csv file with the percent weight of each element in the minderals from the standards:</p>
<pre class="python"><code># import libraries
from sklearn.cluster import DBSCAN as dbscan
from sklearn.linear_model import LinearRegression
import pandas as pd
import numpy as np
from sklearn import metrics
import matplotlib.pyplot as plt
import matplotlib.colors
from sklearn.decomposition import PCA
from pathlib import Path
from skimage.io import imread, imshow
import numpy.ma as ma
from collections import Counter
# read in percent weights by element of the minerals in the standard
weights = pd.read_csv(&quot;/Users/hellenfellows/Desktop/website-hugo/static/mineralmapping/weights_to_minerals.csv&quot;)
print(weights.head())
##        mineral  Mg     Ni  Al        Fe      Ca  Cr   P        S      Ti  Si
## 0  CaTiO3std15 NaN    NaN NaN       NaN  29.481 NaN NaN      NaN  35.211 NaN
## 1  Fe-num2std9 NaN    NaN NaN  100.0000     NaN NaN NaN      NaN     NaN NaN
## 2      FeSstd2 NaN    NaN NaN   63.5252     NaN NaN NaN  36.4748     NaN NaN
## 3   Fe3O4std15 NaN    NaN NaN   72.3591     NaN NaN NaN      NaN     NaN NaN
## 4       Nistd9 NaN  100.0 NaN       NaN     NaN NaN NaN      NaN     NaN NaN</code></pre>
<p>(Note: I was able to add the above Python chunk in R by following <a href="https://babichmorrowc.github.io/post/2019-02-20-pythonrmd/">these instructions</a>).</p>
<p>There is a linear relationship between pixel intensity and percent weight, so we used linear regression to find the slope of this relationship. We read in the .csv of pixel intensities for each element in the standards:</p>
<pre class="python"><code># read in the pixel intensities by element in the standard
mineral_standards = pd.read_csv(&#39;/Users/hellenfellows/Desktop/website-hugo/static/mineralmapping/mineral_standards.csv&#39;)
print(mineral_standards.head())
##    Mg  Ni  Al  Fe   Ca  Cr  P  S   Ti  Si      mineral
## 0   0   0   0   0  171   0  4  0  459   0  CaTiO3std15
## 1   0   0   0   0  148   3  2  0  462   1  CaTiO3std15
## 2   0   2   0   0  141   6  3  0  455   2  CaTiO3std15
## 3   1   2   2   0  122   6  3  0  502   0  CaTiO3std15
## 4   0   0   0   0  138   5  5  0  457   1  CaTiO3std15</code></pre>
<p>We modified the chemical formulas of each mineral using a dictionary by separating each element in the mineral with an “_&quot; to make looping easier.</p>
<pre class="python"><code># create dictionary to standardize file names to chemical formulas
# needed to separate each element in the formula with an _ to make looping easier
mineral_dict = dict(zip(np.unique(mineral_standards[&#39;mineral&#39;]),
    [&quot;Ca_Ti_O_3&quot;, &quot;Fe_&quot;, &quot;Fe_3O_4&quot;, &quot;Fe_S_&quot;, &quot;Ni_S_&quot;, &quot;Ni_&quot;, &quot;Ca_Fe_Mg_Mn_Ni_Si_&quot;, &quot;Ti_O_2&quot;]))
# use dictionary to change mineral columns to underscore format
weights[&#39;mineral&#39;] = weights[&#39;mineral&#39;].map(mineral_dict)
mineral_standards[&#39;mineral&#39;] = mineral_standards[&#39;mineral&#39;].map(mineral_dict)</code></pre>
<p>Next we created a list of the elements accounted for in the standards and made an empty dataframe called <code>coefs</code> to fill with the coefficients of the relationship between pixel intensity and percent weight for each element:</p>
<pre class="python"><code># list of elements
# need to ignore the &quot;mineral&quot; column of the data
elements = [val for val in mineral_standards.columns if val != &#39;mineral&#39;]
coefs = pd.DataFrame(index = [&#39;coeff&#39;], columns = elements)</code></pre>
</div>
<div id="linear-regressions" class="section level3">
<h3>Linear regressions</h3>
<p>Now we looped through the elements to create linear regressions of percent weight vs. pixel intensity based on the intensities in the standards. For these regressions, we forced the intercept to be zero because zero pixel intensity should imply zero percent weight.</p>
<pre class="python"><code># make a linear regression forcing the intercept to be zero
# since zero intensity should correspond to zero percent weight
lr = LinearRegression(fit_intercept = False)
# loop through elements to create linear regression of percent weight vs pixel intensity
# in the minerals in the standard
for element in elements:
    element_df = mineral_standards[mineral_standards[&#39;mineral&#39;].str.contains(element + &quot;_&quot;)]
    # if the element has no percent weights, skip it
    if element_df.empty:
        continue
    minerals = element_df[&#39;mineral&#39;].unique()
    xis = np.empty(0)
    yis = np.empty(0)
    for mine in minerals:
        # get percent weights of the element in that mineral
        weight = weights[weights[&#39;mineral&#39;] == mine][element]
        intensities = element_df[element_df[&#39;mineral&#39;] == mine][element]
        xis = np.append(xis, np.array(intensities))
        yis = np.append(yis, np.repeat(weight, len(intensities)))
    xis, yis = xis.reshape(-1,1), yis.reshape(-1,1)
    # fit linear regression on percent weight vs intensity
    reg = lr.fit(xis,yis)
    xi_pred =  np.arange(0,900).reshape(-1,1)
    # create predictions for range of intensity values
    pred = reg.predict(xi_pred)
    reg.coef_
    # get the linear regression coefficient for each element
    coefs[element] = float(reg.coef_)
# print the coefficients for each element
print(coefs)
##              Mg        Ni   Al        Fe  ...    P         S        Ti        Si
## coeff  0.118599  0.328796  NaN  0.320081  ...  NaN  0.469225  0.077627  0.097906
## 
## [1 rows x 10 columns]</code></pre>
<p>Thus <code>coefs</code> contains the coeficient relating pixel intensity of an element to its percent weight in the mineral: for example, the percent weight of Mg in an mineral is equal to 0.118599 times its pixel intensity.</p>
<figure>
<img class="special-img-class" style="width:100%" src="/img/mineral_regression.png" />
<figcaption>
X-axes are pixel intensity and y-axes are percent weight of the element. Blue points indicate the pixel intensities of the element for a given percent weight based on a mineral in the standard. Orange points indicate the linear regression calculated above.
</figcaption>
</figure>
</div>
<div id="calculate-percent-weights" class="section level3">
<h3>Calculate percent weights</h3>
<p>Finally, we used these coefficients to calculate the predicted percent weights of each element in the two meteorites we were analyzing on a pixel-by-pixel basis. The code for this portion of the analysis is available in the latter end of <a href="https://github.com/HackTheSolarSystem/MineralMapping/blob/master/backend/mineral_mapping_script.py">this script</a>. Note that for any pixel where the percent weight of an element was predicted to be higher than 100%, we set the percent weight to 100%. From these calculations, we ended up with the following files:</p>
<pre class="python"><code># read file of predicted percent weights for meteorite 1
df_obj1 = pd.read_csv(&#39;/Users/hellenfellows/Desktop/website-hugo/static/mineralmapping/predicted_percentweight_obj1.csv&#39;)
df_obj1 = df_obj1.fillna(0)
df_obj1.drop(&#39;Unnamed: 0&#39;, axis = 1, inplace = True)
print(df_obj1.head())
# read file of predicted percent weights for meteorite 2
##          Ca        Ti   Al   Cr  ...    P         Fe        Ni        Mg
## 0  8.595462  0.698642  0.0  0.0  ...  0.0  51.212925  0.986388  5.692734
## 1  4.512618  0.543388  0.0  0.0  ...  0.0  53.453490  1.972775  2.846367
## 2  2.578639  0.465761  0.0  0.0  ...  0.0  63.696075  3.287959  1.067388
## 3  2.148866  0.310508  0.0  0.0  ...  0.0  57.934621  3.945550  0.237197
## 4  1.719092  0.543388  0.0  0.0  ...  0.0  58.254702  3.616755  0.118599
## 
## [5 rows x 10 columns]
df_obj2 = pd.read_csv(&#39;/Users/hellenfellows/Desktop/website-hugo/static/mineralmapping/predicted_percentweight_obj2.csv&#39;)
df_obj2 = df_obj2.fillna(0)
df_obj2.drop(&#39;Unnamed: 0&#39;, axis = 1, inplace = True)
print(df_obj2.head())
##           Si    P   Cr   Al  ...        Ca        Mg        Ni         Fe
## 0  22.322556  0.0  0.0  0.0  ...  0.429773  1.304585  4.931938  14.403635
## 1  11.650808  0.0  0.0  0.0  ...  0.000000  0.355796  2.630367  43.530986
## 2   4.601580  0.0  0.0  0.0  ...  0.429773  0.000000  1.643979  54.093652
## 3   2.349743  0.0  0.0  0.0  ...  0.000000  0.000000  0.986388  57.294460
## 4   0.783248  0.0  0.0  0.0  ...  0.000000  0.000000  1.315183  48.972359
## 
## [5 rows x 10 columns]</code></pre>
<p>Each row in these .csv files corresponds to a pixel in the original image. Each value gives the predicted percent weight of a given element in that pixel.</p>
</div>
</div>
<div id="up-next" class="section level2">
<h2>Up next</h2>
<p>Now that we had converted pixel intensities to predicted percent weights, we were ready to use our clustering algorithm to identify potential minerals in the meteorite. Stay tuned for a future post showing how we used DBSCAN to accomplish this!</p>
</div>
<div id="further-reading" class="section level2">
<h2>Further Reading</h2>
<p>Jeremy Neiman, one of my team members, wrote an excellent <a href="https://towardsdatascience.com/machine-learning-meteorites-fcd8f6859ed7">post</a> describing the challenge in further detail, so check that out for more information.</p>
</div>

    </div>
  
</div>

    </div>
  </section>

  
  
  

  

  

  

  

  

  
  

  
  
  

  
  
  
  
  

  
  

  <section id="2019-02-20-pythonRmd" class="home-section wg-blank   "  >
    <div class="container">
      


<div class="row">
  
    <div class="col-12 col-lg-4 section-heading">
      <h1>Python in RMarkdown</h1>
      
    </div>
    <div class="col-12 col-lg-8">
      


<figure>
<img class="special-img-class" style="width:100%" src="/img/reticulate.png" />
<figcaption>
The <code>reticulate</code> package logo. </figcatption>
</figure>
<div id="using-python-in-rmarkdown" class="section level2">
<h2>Using Python in RMarkdown</h2>
<p>In order to write blog posts using Python code, I wanted to figure out a way to include Python code chunks in RMarkdowns. When you insert a code chunk in RMarkdown, you have the option of specifying the language of that chunk: the default is R, but you can also insert a Bash, SQL, Python, etc. code chunk.</p>
<p>When I attempted to insert a Python code chunk and import libraries, however, I kept getting the error:</p>
<p><code>Error in py_run_string_impl(code, local, convert) :</code> <code>ImportError: No module named sklearn.cluster</code></p>
<p>From running Python in Atom, I knew I had the <code>sklearn.cluster</code> module installed, so the problem must be in the connection between R and Python.</p>
</div>
<div id="reticulate" class="section level2">
<h2><code>reticulate</code></h2>
<p>The <code>reticulate</code> package in R (website <a href="https://rstudio.github.io/reticulate/index.html">here</a> allows R to interact with Python. I installed the package from RStudio.</p>
<pre class="r"><code># install.packages(&quot;reticulate&quot;)
library(reticulate)</code></pre>
</div>
<div id="changing-python-versions" class="section level2">
<h2>Changing Python versions</h2>
<p>Installing <code>reticulate</code> still didn’t allow me to knit the RMarkdown with a Python code chunk, however. I followed the instructions in <a href="https://rstudio-pubs-static.s3.amazonaws.com/397064_bd288708515c4b71a6e78dfe91b8bb99.html">this post</a> by Pablo Franco to check the Python version that <code>reticulate</code> was using:</p>
<pre class="r"><code>py_discover_config()</code></pre>
<p>I ended up with the following output:</p>
<pre><code>python:         /usr/bin/python
libpython:      /System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/config/libpython2.7.dylib
pythonhome:     /System/Library/Frameworks/Python.framework/Versions/2.7:/System/Library/Frameworks/Python.framework/Versions/2.7
version:        2.7.10 (default, Aug 17 2018, 19:45:58)  [GCC 4.2.1 Compatible Apple LLVM 10.0.0 (clang-1000.0.42)]
numpy:          /System/Library/Frameworks/Python.framework/Versions/2.7/Extras/lib/python/numpy
numpy_version:  1.8.0</code></pre>
<p>I wanted to be running Python version 3.6, which was the version I had installed using Anaconda, so I needed to change the path.</p>
<div id="set-up-chunk" class="section level3">
<h3>Set-up chunk</h3>
<p>I discovered that you can set the path to a different installation of Python by modifying the setup chunk at the start of the RMarkdown. According to the <a href="https://bookdown.org/yihui/rmarkdown/language-engines.html"><code>bookdown</code> website</a>, the default used is Python 2.</p>
<p>My default version of this set-up chunk looks like this:</p>
<pre><code>{r setup, include=FALSE}
knitr::opts_chunk$set(collapse = TRUE)</code></pre>
<p>You can set the chunk option <code>engine.path</code> to specify the path to the engine interpreter and change it from the default Python 2.</p>
</div>
<div id="finding-python-path" class="section level3">
<h3>Finding Python path</h3>
<p>I now needed to find the actual path to Python that I wanted to use. I did this by opening up Python separately from RStudio (I used Atom for this) and running the following (I got the code for this from <a href="https://www.dummies.com/programming/python/how-to-find-path-information-in-python/">here</a>):</p>
<pre class="python"><code>import sys
for p in sys.path:
    print(p)
## /Applications/anaconda3/bin
## /Applications/anaconda3/lib/python37.zip
## /Applications/anaconda3/lib/python3.7
## /Applications/anaconda3/lib/python3.7/lib-dynload
## /Applications/anaconda3/lib/python3.7/site-packages
## /Applications/anaconda3/lib/python3.7/site-packages/aeosa
## /Library/Frameworks/R.framework/Versions/3.5/Resources/library/reticulate/python</code></pre>
<p>From this information, I could tell I wanted to use the path <code>/anaconda3/lib/python3.6</code>, rather than <code>/usr/bin/python</code>, which is what RMarkdown had originally been using. I modified by set-up chunk to look like this:</p>
<pre><code>{r setup, include=FALSE}
knitr::opts_chunk$set(collapse = TRUE, engine.path = list(python = &#39;/anaconda3/bin/python3.6&#39;))</code></pre>
</div>
</div>
<div id="other-options" class="section level2">
<h2>Other options</h2>
<p>This solution enabled me to knit RMarkdowns with Python code chunks! It changes the engine interpreter globally, which you could do for multiple engines simultaneously, like Python and Ruby, for example:</p>
<pre><code>knitr::opts_chunk$set(engine.path = list(
  python = &#39;/anaconda3/bin/python3.6&#39;,
  ruby = &#39;/usr/local/bin/ruby&#39;
))</code></pre>
<p>Alternatively, you can specify the engine interpreter locally in each code chunk by starting the chunk with <code>{python, engine.path = '/anaconda3/bin/python3.6</code>}, for example.</p>
</div>

    </div>
  
</div>

    </div>
  </section>

  
  
  

  

  

  

  

  

  
  

  
  
  

  
  
  
  
  

  
  

  <section id="2019-02-14-hackathon" class="home-section wg-blank   "  >
    <div class="container">
      


<div class="row">
  
    <div class="col-12 col-lg-4 section-heading">
      <h1>#HackTheSolarSystem</h1>
      
    </div>
    <div class="col-12 col-lg-8">
      


<figure>
<img class="special-img-class" style="width:40%" src="/img/hackthesolarsystem.png" />
</figure>
<p>Last weekend I had the opportunity to participate in my first ever hackathon: the #HackTheSolarSystem hackathon hosted by the American Museum of Natural History.</p>
<div id="what-is-a-hackathon-anyways" class="section level2">
<h2>What is a hackathon, anyways?</h2>
<p>This is a question I got asked countless times this last week as I informed my friends and family that I would be at work <strong>all</strong> weekend (yes, including Friday night, yes, including Saturday night, no, I will not be able to socialize at all). And, in fairness, this is a question I wouldn’t have had an answer to prior to this job. A hackathon is an event where participants come together to use technology to address a challenge. Usually, the event takes place over a few days, where hackers form teams and collaborate on their projects, before presenting their solutions.</p>
<p>When I learned about this concept, I was baffled by the idea that it would be possible to come up with any kind of viable solution to a problem in only 24 hours. Coming from an academic background, I’m used to months of reading papers to get oriented to a problem before I can produce any kind of meaningful code. This event forced us to dive right in, looking at the problem from a computer science standpoint, rather than striving to get the full scientific background on the problem. I was amazed by the ability of all the teams at the event to create meaningful solutions to scientific problems they had not necessarily studied before.</p>
</div>
<div id="the-ammh-hackathon" class="section level2">
<h2>The AMMH Hackathon</h2>
<p>For the past five years, the American Museum of Natural History has hosted a hackathon. #HackTheSolarSystem is the first and only hackathon I have participated in thus far, so I can’t compare it to other events, but my understanding is that the museum seeks to put on a slightly different type of event than other hackathons: AMNH wants to give technologists and scientists the opportunity to work together to create solutions to real research problems. For #HackTheSolarSystem, museum scientists from the Department of Earth and Planetary Science put forth a series of challenges they face in their research that they thought could be solved through technology. These challenges ranged from tracking dust particles through aerogel to visualizing the Sun (a full list of challenges is available <a href="https://github.com/amnh/HackTheSolarSystem/wiki/Challenges">here</a>). In addition to the scientific challenges, there were also educational challenges for both high school students and educators.</p>
<p>On Friday night, the museum “stakeholders”, i.e. the scientists posing the challenges, presented their challenges and led participants on tours of various areas of the museum, like the Hall of Meteorites and the Earth and Planetary Sciences department. Then, from Saturday at 2:00 pm to Sunday at 1:00 pm, us hackers stayed at the museum to develop solutions to our chosen challenges. Most hackers worked on teams formed at the event, although some corporate teams came into the hackathon with a team already formed.</p>
</div>
<div id="the-challenge" class="section level2">
<h2>The Challenge</h2>
<p>I chose to work on the meteorite mineral mapping challenge: our “stakeholders”, museum scientists Marina Gemma and Sam Alpert, wanted a way to identify mineral composition in images of meteorites. Each meteorite slice had been imaged, resulting in data consisting of pixel intensity for a handful of elements. We also had access to a series of standards with known mineral content so we could compare the pixel intensities in these known minerals with the pixel intensities in the unknown minerals.</p>
<p>Our team (pictured below) used four different approaches to address this problem: linear classification via SVM, random forest classification, nearest neighbor classification, and cluster inference. You can check out more details on these solutions in the Readme of our GitHub repository: <a href="https://github.com/HackTheSolarSystem/MineralMapping" class="uri">https://github.com/HackTheSolarSystem/MineralMapping</a>. (And I will post a more detailed explanation of the clustering solution in a later post, which is the part of the code I focused on!)</p>
<figure>
<img class="special-img-class" style="width:100%" src="/img/hackathon_team.jpg" />
<figcaption>
From left to right: Peter Kang, Jackson Lee, Jeremy Neiman, John Underwood, Katy Abbott, Cecina Babich Morrow, Meret Götschel
</figcaption>
</figure>
<p>Our solution ended up making us one of the four primary winning teams at the hackathon! We won the “LabCoat Knockout” award for creating a solution with the potential to change the nature of research. We plan to continue working with our museum stakeholders to fine-tune our code so that it can be used in their research in the future. To me, that is the most exciting part of this event - that a group of technologists from diverse backgrounds (none of which knew much about meteorites before this event!) could write code in 24 hours with the potential to solve a scientific problem with a concrete impact on research at the museum.</p>
</div>

    </div>
  
</div>

    </div>
  </section>

  
  
  

  

  

  

  

  

  
  

  
  
  

  
  
  
  
  

  
  

  <section id="2019-01-09-grafting-trees" class="home-section wg-blank   "  >
    <div class="container">
      


<div class="row">
  
    <div class="col-12 col-lg-4 section-heading">
      <h1>Grafting phylogenies</h1>
      
    </div>
    <div class="col-12 col-lg-8">
      


<figure>
<img class="special-img-class" style="width:100%" src="/img/bodymass_tetrapod_tree.png" />
</figure>
<div id="inspiration-for-this-post" class="section level1">
<h1>Inspiration for this post</h1>
<p>This post comes from the finishing touches I needed to put on a paper about <a href="https://babichmorrowc.github.io/project/life_history/">life history evolution</a>. The paper compares life history traits across the four groups of tetrapods (amphibians, reptiles, mammals, and birds), so when I carried out the nitty gritty phylogenetic analyses, I used four separate phylogenies. When it came time to make a figure to visualize those analyses, however, I ended up with an unwieldy (and unpublishable) 16-panel figure (four phylogenies by four traits). One of my coauthors suggested using a tetrapod supertree to visualize the evolution of the traits across all four classes simultaneously.</p>
<div id="uyeda-et-al.-2017" class="section level2">
<h2>Uyeda et al. 2017</h2>
<p>Uyeda et al. did something similar in their 2017 paper <a href="https://www.journals.uchicago.edu/doi/10.1086/692326">The evolution of energetic scaling across the vertebrate tree of life</a>. They stitched together fish, amphibian, squamate, bird, and mammal phylogenies together to visualize metabolic rate across all vertebrates:</p>
<figure>
<img class="special-img-class" style="width:75%" src="/img/uyeda_figure.png" />
<figcaption>
Figure 1 from Uyeda et al. 2017
</figcaption>
</figure>
<p>At first, I was very hopeful that I would be able to download this supertree and prune it to the taxa in my analysis since the authors were using the same clade specific phylogenies that I was. The phylogeny is available on Data Dryad (<a href="https://datadryad.org/resource/doi:10.5061/dryad.3c6d2" class="uri">https://datadryad.org/resource/doi:10.5061/dryad.3c6d2</a>). Unfortunately, after downloading that phylogeny and pruning it to include species I used in my analysis, I ended up with approximately 15% of the species I analyzed in the resulting tree.</p>
</div>
</div>
<div id="into-the-code" class="section level1">
<h1>Into the code</h1>
<p>Since my easy solution didn’t pan out and I couldn’t get enough information from the supplemental material for the paper to replicate their analyses, I looked on GitHub to try to find Uyeda’s code. Hooray for GitHub once again, because the repository for the paper can be found here: <a href="https://github.com/uyedaj/bmr" class="uri">https://github.com/uyedaj/bmr</a>. The <a href="https://github.com/uyedaj/bmr/blob/master/R/ManuscriptNotebook.Rmd">RMarkdown</a> details the analyses for the paper, including the process for making the full tree.</p>
<p>Start with loading the necessary packages.</p>
<pre class="r"><code>library(phytools)
## Loading required package: ape
## Warning: package &#39;ape&#39; was built under R version 3.5.2
## Loading required package: maps
library(geiger)</code></pre>
<div id="the-original-phylogenies" class="section level2">
<h2>The original phylogenies</h2>
<p>I was working with four separate phylogenies: amphibians, squamates, birds, and mammals. For amphibians, I used a congruified time-tree from the <code>PhyloOrchard</code> package (O’Meara et al. 2013) that was constructed using the Alfaro et al. timetree of gnathostomes (Alfaro et al. 2009) as the reference and the Pyron and Wiens amphibian phylogeny as the target (Pyron and Wiens 2011).</p>
<pre class="r"><code>length(amphibiantree$tip.label)
## [1] 2871
plot(amphibiantree, type = &quot;fan&quot;, show.tip.label = FALSE)</code></pre>
<p><img src="/post/2019-01-09-grafting-trees_files/figure-html/unnamed-chunk-3-1.png" width="672" /></p>
<p>For squamates, I used the Zheng and Wiens time-calibrated phylogeny (Zheng and Wiens 2016).</p>
<pre class="r"><code>length(squamatetree$tip.label)
## [1] 378
plot(squamatetree, type = &quot;fan&quot;, show.tip.label = FALSE)</code></pre>
<p><img src="/post/2019-01-09-grafting-trees_files/figure-html/unnamed-chunk-4-1.png" width="672" /></p>
<p>For birds, I used the Jetz phylogeny (Jetz et al. 2012).</p>
<pre class="r"><code>length(birdtree$tip.label)
## [1] 9993
plot(birdtree, type = &quot;fan&quot;, show.tip.label = FALSE)</code></pre>
<p><img src="/post/2019-01-09-grafting-trees_files/figure-html/unnamed-chunk-5-1.png" width="672" /></p>
<p>For mammals, I used the supertree from Fritz et al. 2009.</p>
<pre class="r"><code>length(mammaltree$tip.label)
## [1] 5020
plot(mammaltree, type = &quot;fan&quot;, show.tip.label = FALSE)</code></pre>
<p><img src="/post/2019-01-09-grafting-trees_files/figure-html/unnamed-chunk-6-1.png" width="672" /></p>
</div>
<div id="reading-in-trees" class="section level2">
<h2>Reading in trees</h2>
<p>The first step of the process is reading in the individual phylogenies you want to stitch together. This step is straightforward, with one exception: you cannot have species that are present in multiple of the individual trees. For example, my squamate phylogeny included <em>Gallus gallus</em> (red junglefowl) and <em>Dromaius novaehollandia</em> (emu). Since these species were also present in my bird phylogeny, I got the following error: <code>Found matching tips in 'subtree' and 'phy'</code>. To solve this problem, I just removed these tips from the squamate tree:</p>
<pre class="r"><code>squamatetree &lt;- drop.tip(phy = squamatetree, tip = c(&quot;Gallus_gallus&quot;, &quot;Dromaius_novaehollandiae&quot;))</code></pre>
<p>I ended up with the following list of trees and corresponding tip labels:</p>
<pre class="r"><code>tree_list &lt;- list(amphib=amphibiantree, birds=birdtree, squam=squamatetree, mamm=mammaltree)
class(tree_list) &lt;- &quot;multiPhylo&quot;</code></pre>
</div>
<div id="make-a-tree-with-orders" class="section level2">
<h2>Make a tree with orders</h2>
<p>In Uyeda et al. (2017), the authors were creating a phylogeny for all vertebrates, but for my analyses I was only examining tetrapods, so I didn’t have a fish phylogeny to include. The original code from Uyeda et al. to create a tree with the 5 vertebrate orders looks like this:</p>
<pre class="r"><code>tip.labels &lt;- c(&quot;fish&quot;, &quot;amphib&quot;, &quot;squam&quot;, &quot;birds&quot;, &quot;mamm&quot;)

## Make a tree with just orders:
edge &lt;- matrix(c(9, 4,
  9, 3,
  8, 5,
  8, 9,
  7, 8,
  7, 2,
  6, 7,
  6, 1), byrow=TRUE, ncol=2)
## Dates from Timetree of life (timetree.org)
edge.length &lt;- c(274.9, 274.9, 324.5, 324.5-274.9, 382.9-324.5, 382.9, 454.6-382.9 , 454.6)
Nnode &lt;- 4
ordertree &lt;- list(edge=edge, Nnode=Nnode, tip.label=tip.labels, edge.length=edge.length)
class(ordertree) &lt;- &#39;phylo&#39;
plot(ordertree)</code></pre>
<p><img src="/post/2019-01-09-grafting-trees_files/figure-html/unnamed-chunk-9-1.png" width="672" /></p>
<p>To visualize the results, we can add tip labels, node labels, and edge labels to the tree with the branch lengths:</p>
<pre class="r"><code>plot(ordertree)
tiplabels()
nodelabels()
edgelabels(ordertree$edge.length, bg=&quot;black&quot;, col=&quot;white&quot;, font=2)</code></pre>
<p><img src="/post/2019-01-09-grafting-trees_files/figure-html/unnamed-chunk-10-1.png" width="672" /></p>
<div id="getting-rid-of-fish" class="section level3">
<h3>Getting rid of fish</h3>
<p>Since I didn’t have fish, I needed to make a few modifications. First, <code>tip.labels</code> didn’t need “fish” in it anymore:</p>
<pre class="r"><code># remove &quot;fish&quot; from tip.labels:
tip.labels &lt;- c(&quot;amphib&quot;, &quot;squam&quot;, &quot;birds&quot;, &quot;mamm&quot;)</code></pre>
<p>Now, for the trickier part - I needed to modify the edge matrix. The edge matrix contains the starting and ending nodes for each edge in a tree. As we can see from the plot above, numbering works in the following way: the tips are numbered starting at the top from 1 to the number of tips and the nodes are numbered starting at the root and moving towards the tips. To get rid of fish, I needed to delete one tip from the tree and one node (the original root node). I sketched out what I wanted the new order tree to look like, complete with numbered nodes and tips, and created the following edge matrix:</p>
<pre class="r"><code>edge &lt;- matrix(c(7, 3,
  7, 2,
  6, 4,
  6, 7,
  5, 6,
  5, 1), byrow=TRUE, ncol=2)</code></pre>
<p>Since I was losing two edges from the phylogeny (the one going from the root to fish and the one from the root to the last common ancestor of tetrapods), I also needed to modify the edge lengths by removing 454.6-382.9 and 454.6:</p>
<pre class="r"><code>edge.length &lt;- c(274.9, 274.9, 324.5, 324.5-274.9, 382.9-324.5, 382.9)</code></pre>
<p>The final modification I needed was to decrease <code>Nnode</code> from 4 to 3:</p>
<pre class="r"><code>Nnode &lt;- 3</code></pre>
<p>So now…</p>
<pre class="r"><code>ordertree &lt;- list(edge=edge, Nnode=Nnode, tip.label=tip.labels, edge.length=edge.length)
class(ordertree) &lt;- &#39;phylo&#39;
plot(ordertree)
edgelabels(ordertree$edge.length, bg=&quot;black&quot;, col=&quot;white&quot;, font=2)</code></pre>
<p><img src="/post/2019-01-09-grafting-trees_files/figure-html/unnamed-chunk-15-1.png" width="672" /></p>
<p>…I was ready to go with an order-level tree onto which I could graft my individual phylogenies!</p>
</div>
<div id="node-dates" class="section level3">
<h3>Node dates</h3>
<p>…Not so fast. I ended up with an additional problem I needed to solve before grafting the trees together. I ran into the error <code>'split_age' is inconsistent with edge lengths in 'phy'</code>, which means that the earliest node in one of my individual phylogenies was older than the node age I gave in <code>edge.length</code>. By <a href="https://babichmorrowc.github.io/post/debug-r/">using <code>debug</code></a>, I was able to tell that the error occurred when I added the squamate tree. The oldest node in my squamate tree was 277.8 million years ago, but I had set the divergence time between birds and squamates at 274.9 mya, so R was having problems. The species causing the problem was the tuatara, which is the only surviving member of its order.</p>
<figure>
<img class="special-img-class" style="width:75%" src="/img/tuatara.png" />
<figcaption>
The pesky (yet very cute) tuatara (<a href="https://www.australiangeographic.com.au/blogs/creatura-blog/2017/12/the-tuatara/" class="uri">https://www.australiangeographic.com.au/blogs/creatura-blog/2017/12/the-tuatara/</a>)
</figcaption>
</figure>
<p>I had a couple of choices: either delete the tuatara from the squamate phylogeny or increase the age of the last common ancestor of birds and squamates when I created the vector <code>edge.length</code>. I chose to do the latter because why get rid of such a cool animal!</p>
<p>I went to <a href="http://timetree.org/" class="uri">http://timetree.org/</a> to see if I could find a reasonable range of estimates for this node. According to the website, which allows you to search for the divergence time between any two taxa, the estimated divergence of birds and squamates occurred 280 mya.</p>
<figure>
<img class="special-img-class" style="width:75%" src="/img/timetree_birdsquamate.png" />
<figcaption>
TimeTree results for divergence time of birds and squamates.
</figcaption>
</figure>
<p>So I ended up with the following code and order tree:</p>
<pre class="r"><code>tip.labels &lt;- c(&quot;amphib&quot;, &quot;squam&quot;, &quot;birds&quot;, &quot;mamm&quot;)
edge &lt;- matrix(c(7, 3,
  7, 2,
  6, 4,
  6, 7,
  5, 6,
  5, 1), byrow=TRUE, ncol=2)
edge.length &lt;- c(280, 280, 324.5, 324.5-274.9, 382.9-324.5, 382.9)
Nnode &lt;- 3
ordertree &lt;- list(edge=edge, Nnode=Nnode, tip.label=tip.labels, edge.length=edge.length)
class(ordertree) &lt;- &#39;phylo&#39;
plot(ordertree)
edgelabels(ordertree$edge.length, bg=&quot;black&quot;, col=&quot;white&quot;, font=2)</code></pre>
<p><img src="/post/2019-01-09-grafting-trees_files/figure-html/unnamed-chunk-16-1.png" width="672" /></p>
<pre class="r"><code>tree_list &lt;- list(amphib=amphibiantree, birds=birdtree, squam=squamatetree, mamm=mammaltree)
class(tree_list) &lt;- &quot;multiPhylo&quot;</code></pre>
</div>
</div>
<div id="grafting-the-trees" class="section level2">
<h2>Grafting the trees</h2>
<p>The final step is grafting the individual trees onto the order tree in the proper place.</p>
<pre class="r"><code>#Add taxonomic information to tree
otax &lt;- data.frame(&quot;Class&quot;= ordertree$tip.label, &quot;Superclass&quot;=c(rep(&quot;Tetrapoda&quot;,2)))
rownames(otax) &lt;- ordertree$tip.label
classtree &lt;- nodelabel.phylo(ordertree, otax, ncores=1)

res &lt;- glomogram.phylo(classtree, tree_list)
plot(res, type = &quot;fan&quot;, show.tip.label = FALSE)</code></pre>
<p><img src="/post/2019-01-09-grafting-trees_files/figure-html/unnamed-chunk-18-1.png" width="672" /></p>
<p>Voila - a tree with 18262 species of tetrapods!</p>
</div>
</div>
<div id="disclaimer" class="section level1">
<h1>Disclaimer</h1>
<p>I’d like to finish this post with a disclaimer: I am NOT a phylogeneticist (yet?). The supertree created in this analysis incorporates several different phylogenies from literature and adapts code from another published article (all written by people with much more phylogenetic background than I!). However, the accuracy of the tree decreases as you move back in time - there is a great deal of uncertainty about node age for the deeper nodes in the tree. Even so, this process allows us to make some cool visualizations to compare major clades across vast stretches of evolutionary time - even if precise dates are incorrect, overall patterns are still informative!</p>
</div>
<div id="code" class="section level1">
<h1>Code</h1>
<p>The entire script I used for this process can be found at <a href="https://github.com/KerkhoffLab/bodymasspatterns/blob/master/tetrapod_phylogeny_code.R" class="uri">https://github.com/KerkhoffLab/bodymasspatterns/blob/master/tetrapod_phylogeny_code.R</a>.</p>
</div>
<div id="literature-cited" class="section level1">
<h1>Literature Cited</h1>
<p>Alfaro, M. E., F. Santini, C. Brock, H. Alamillo, A. Dornburg, D. L. Rabosky, G. Carnevale, and L. J. Harmon. 2009. Nine exceptional radiations plus high turnover explain species diversity in jawed vertebrates. PNAS 106:13410-13414.</p>
<p>Fritz, S. A., O. R. P. Bininda-Emonds, and A. Purvis. 2009. Geographical variation in predictors of mammalian extinction risk: big is bad, but only in the tropics. Ecology Letters 12:538–549.</p>
<p>Jetz, W., G. H. Thomas, J. B. Joy, K. Hartmann, and A. O. Mooers. 2012. The global diversity of birds in space and time. Nature 491:444.</p>
<p>O’Meara, B. C., L. J. Harmon, and J. Eastman. 2013. PhyloOrchard: Important and/or useful phylogenetic datasets.</p>
<p>Pyron, R. A. and J. J. Wiens. 2011. A large-scale phylogeny of Amphibia including over 2800 species, and a revised classification of extant frogs, salamanders, and caecilians. Molecular Phylogenetics and Evolution 61: 543-583.</p>
<p>Uyeda JC, Pennell MW, Miller ET, Maia R, McClain CR (2017) The evolution of energetic scaling across the vertebrate tree of life. The American Naturalist 190(2): 185-199. <a href="https://doi.org/10.1086/692326" class="uri">https://doi.org/10.1086/692326</a></p>
<p>Uyeda JC, Pennell MW, Miller ET, Maia R, McClain CR (2017) Data from: The evolution of energetic scaling across the vertebrate tree of life. Dryad Digital Repository. <a href="https://doi.org/10.5061/dryad.3c6d2" class="uri">https://doi.org/10.5061/dryad.3c6d2</a></p>
<p>Zheng, Y., and J. J. Wiens. 2016. Combining phylogenomic and supermatrix approaches, and a time-calibrated phylogeny for squamate reptiles (lizards and snakes) based on 52 genes and 4162 species. Molecular Phylogenetics and Evolution 94:537–547.</p>
</div>

    </div>
  
</div>

    </div>
  </section>

  
  
  

  

  

  

  

  

  
  

  
  
  

  
  
  
  
  

  
  

  <section id="2019-01-03-debugging-R" class="home-section wg-blank   "  >
    <div class="container">
      


<div class="row">
  
    <div class="col-12 col-lg-4 section-heading">
      <h1>Debugging in R</h1>
      
    </div>
    <div class="col-12 col-lg-8">
      


<figure>
<img class="special-img-class" style="width:75%" src="/img/PhD_Debugging.png" />
<figcaption>
PhD comics Jorge Cham (www.phdcomics.com)
</figcaption>
</figure>
<p>Recently I’ve been part of beta testing a new R package. The package allows users to apply different masks to spcies distribution models to create more ecologically realistic range models. The package itself is very cool, but since it hasn’t been released yet, this post will be short on ecology. Instead, I’m going to focus on the process I’ve been using to debug the package. This debugging thought process and the different functions that make it possible are things I wish I’d been formally taught while learning R (but better late than never!).</p>
<div id="traceback" class="section level1">
<h1><code>traceback</code></h1>
<p>The <code>traceback()</code> function is definitely one of those that I wished I learned sooner. Basically, <code>traceback()</code> shows you the series of functions called, including the one resulting in the last error. For example, when I ran <code>traceback()</code> after the function I was testing threw an error, I got the following readout:</p>
<figure>
<img class="special-img-class" style="width:100%" src="/img/traceback_example.png" />
</figure>
<p>When I tried to use the function <code>rangeSVM()</code>, I got an error, as we can see in the first line of code I ran. By running <code>traceback()</code>, I can see the order of functions that <code>rangeSVM()</code> uses: first, it called the <code>svm()</code> function from the R package <code>e1071</code>, which in turn called <code>svm.formula()</code>, which finally called <code>svm.default()</code>. This last function, <code>svm.default()</code> is what produced the error (as we can see from the error message, as well).</p>
<p>Running <code>traceback()</code> can be especially helpful in situations like this, where the error message comes from a function called internally by the function you actually ran. Here, the error does not come from the <code>rangeSVM()</code> function, so reading the documentation for <code>rangeSVM()</code> could not save me. Now I know that the error actually comes from a function in the <code>e1071</code> package, so I could check out the source code to figure out what kind of issue causes that error message.</p>
</div>
<div id="checking-source-code" class="section level1">
<h1>Checking source code</h1>
<p>There is a <a href="https://github.com/cran">CRAN GitHub account</a> that is an unofficial read-only mirror of all CRAN packages. Typically, when I want to read through the code for a particular function in a package, this is what I use. To dig a little deeper into what was going wrong with the <code>svm()</code> function, I found the code <a href="https://github.com/cran/e1071/blob/master/R/svm.R">here</a>. This allowed me to look at the function definitions for <code>svm.formula()</code> as well as <code>svm.default()</code>.</p>
</div>
<div id="debug" class="section level1">
<h1><code>debug</code></h1>
<p>Reading through the source code is helpful, but wouldn’t it be even better if you could watch R step through the function line by line until the error? Fortunately the handy <code>debug()</code> function does exactly that. By running the following:</p>
<pre class="r"><code>debug(rangeSVM)
svmHYB_weight &lt;- rangeSVM(variegatus[,2:3], tridactylus[,2:3], sdm = raster::stack(var_sdm, tri_sdm), nrep = 3, weight = TRUE)</code></pre>
<p>I could see that the error occurred at line 100 of <code>svm.R</code>:</p>
<pre class="r"><code># Browse[2]&gt; 
# debug at /Users/hellenfellows/OneDrive - AMNH/Wallace/maskRangerBetaTesting/maskRangeR/R/svm.R#100: m &lt;- e1071::svm(sp ~ ., data = xy, gamma = params_best_df_mostFreq$gamma[1], 
#     cost = params_best_df_mostFreq$cost[1], class.weights = cw)
# Browse[2]&gt; 
# Error in svm.default(x, y, scale = scale, ..., na.action = na.action) : 
#   NA/NaN/Inf in foreign function call (arg 10)</code></pre>
<p>The <code>Browse&gt;</code> prompt at the beginning of the lines indicate that the debugger is working. The <code>debug at</code> part shows the next line of code to be executed when you hit enter next. R will continue to show you the following line of code as you press enter (up until you hit the error). If you are running all of this in RStudio, it will also conveniently highlight the lines in the source code that will be run next. Also, when you want to get out of the debugger so you can run that function again later without debugging, run <code>undebug(rangeSVM)</code>, for example.</p>
<p>My initial suspicion was that the function was failing during one of the runs of a for loop, however the debugger showed me that the function was making it through all the runs of the for loop. Instead, the error occurred during the final step of <code>rangeSVM()</code> when the function attempted to integrate the results of all runs of the for loop to create a final support vector machine model.</p>
<div id="get" class="section level2">
<h2><code>get</code></h2>
<p>To figure out why this problem was occurring, I wanted to be able to see the values of the different variables passeed to that final step of the function. Handily enough, while in the debugger, you can use the function <code>get(&quot;variable&quot;)</code> to check the status of the different variables (the name of the variable needs to be in quotation marks).</p>
<p>For example:</p>
<pre class="r"><code>debug(rangeSVM)
svmHYB &lt;- rangeSVM(variegatus[,2:3], tridactylus[,2:3], sdm = raster::stack(var_sdm, tri_sdm), nrep = 3)
# Browse[2]&gt; 
# debug at /Users/hellenfellows/OneDrive - AMNH/Wallace/maskRangerBetaTesting/maskRangeR/R/svm.R#96: params_best_df$params &lt;- paste0(params_best_df$gamma, params_best_df$cost)

# Browse[2]&gt; get(&quot;params_best_df&quot;)
#        gamma cost class.weights
# 38 0.5000000    2             1
# 85 0.0078125 2048             1
# 95 0.0078125 8192             1</code></pre>
<p>This was the final step of the process that actually allowed me to figure out the problem:</p>
<pre class="r"><code># Browse[2]&gt; get(&quot;params_best_df_mostFreq&quot;)
# [1] gamma cost 
# &lt;0 rows&gt; (or 0-length row.names)</code></pre>
<p>Without going into too much detail about the function itself, I was able to tell that the parameters I was feeding into the support vector machine function somehow didn’t exist: the variable <code>params_best_df_mostFreq</code> had 0 rows.</p>
</div>
</div>
<div id="other-tips-and-tricks" class="section level1">
<h1>Other tips and tricks</h1>
<p>I got much of these resources from a very helpful blog post called <a href="https://petewerner.blogspot.com/2013/01/tracking-down-errors-in-r.html">“Tracking down errors in R”</a> by Pete Werner (also available as a post on <a href="https://www.r-bloggers.com/tracking-down-errors-in-r/">R-bloggers</a>). This post goes through a slightly simpler example of debugging which is reproducible (something I have not attempted to do here), so it is very helpful for seeing the process on a simpler function. In addition to the techniques I used, Pete also explains how you can turn warning messages into errors in the case that your function is throwing a troubling warning message that you suspect is causing an error later on.</p>
</div>

    </div>
  
</div>

    </div>
  </section>

  
  
  

  

  

  

  

  

  
  

  
  
  

  
  
  
  
  

  
  

  <section id="2018-12-12-add-google-doc" class="home-section wg-blank   "  >
    <div class="container">
      


<div class="row">
  
    <div class="col-12 col-lg-4 section-heading">
      <h1>Adding Google Docs to website</h1>
      
    </div>
    <div class="col-12 col-lg-8">
      


<p><img class="special-img-class" style="width:75%" src="/img/google_docs_fig.png" /></p>
<p>Given that the Academic website theme comes with a place to upload talks, I thought I might as well upload the slides for my Senior Honors presentation. Unfortunately, this was a slightly more difficult task than I anticipated. The default format for slides in Hugo appears to be Markdown slides, which is fine, except when your talk was made in Google Slides.</p>
<div id="google-slides-to-markdown" class="section level2">
<h2>Google Slides to Markdown?</h2>
<p>My first thought was that there must be some tool to convert a Google slides talk into a Markdown document. For whatever reason, there are several tools (like <a href="https://github.com/gsuitedevs/md2googleslides">this GitHub repo</a>) to go the other direction, from Markdown to Google slides. I found a GitHub repo <a href="https://github.com/mangini/gdocs2md"><code>gdocs2md</code></a> that claimed to go in the other direction, but the script provided doesn’t currently run (if anyone is a whiz at Google Apps and wants to take a stab at fixing up this code, it would be a great tool!). There is a Google Drive add-on called <a href="https://chrome.google.com/webstore/detail/gd2md-html/igffnbdfnodiaphfmfaiiaegmoljbghf">GD2md-html</a> that converts Google Docs to Markdown or HTML, which should be useful for the future, but it doesn’t help with Google Slides.</p>
</div>
<div id="embedding-google-slides-in-hugo" class="section level2">
<h2>Embedding Google Slides in Hugo</h2>
<p>Having abandoned the idea of just converting my current slides into Markdown and going on my merry way, I started looking for a way to embed the Google Slides themselves into my website.</p>
<div id="hugo-shortcodes" class="section level3">
<h3>Hugo shortcodes</h3>
<p>Hugo has certain snippets called <a href="https://gohugo.io/content-management/shortcodes/">shortcodes</a> that are designed to address the problem of having to add raw HTML into Markdown. These shortcodes help render a content file according to a template created by Hugo. For example, the shortcode <code>figure</code> adds HTML functionality to the Markdown image syntax. There are shortcodes for Instagram (<code>instagram</code>), Vimeo (<code>vimeo</code>) and Twitter (<code>tweet</code>).</p>
</div>
<div id="the-gdocs-shortcode" class="section level3">
<h3>The <code>gdocs</code> shortcode</h3>
<p>One of these shortcodes, <code>gdocs</code>, allows you to embed all Google doc types (Slides, Docs, Sheets, etc.) into your website. The HTML code for the shortcode can be found in the <code>hugo-academic</code> GitHub repo: <a href="https://github.com/gcushen/hugo-academic/blob/master/layouts/shortcodes/gdocs.html" class="uri">https://github.com/gcushen/hugo-academic/blob/master/layouts/shortcodes/gdocs.html</a>.</p>
<p>I tried inserting the following into the <code>index.md</code> file for my talk below the line with three <code>+</code> signs:</p>
<pre class="bash"><code>{{&lt; gdocs &quot;https://docs.google.com/presentation/d/e/2PACX-1vTsSj-oftLksGEKTkzCoBSIpzooO61ZMGAUzoGBv7u20OzMQK8ctX3gCpWTMm4UowrrWumFrdQsY0Cd/embed?start=false&amp;loop=false&amp;delayms=5000&quot; &gt;}}</code></pre>
<p>I got the URL by going to my Google Slides, and clicking File &gt; Publish to the web. I selected Embed and copied the link following <code>src=</code>.</p>
<p>Unfortunately, this didn’t result in anything showing up on my page besides a large blank space where I assume Hugo was trying to put my presentation. I tried some variations on the above code, including adding <code>src=</code> before the URL (no change), and moving the entire code snippet above the <code>+++</code> line in <code>index.md</code> (results in the error &quot; unmarshal failed: Near line 65 (last key parsed ‘image’): bare keys cannot contain ‘{’ “).</p>
</div>
<div id="html-code" class="section level3">
<h3>HTML code</h3>
<p>At this point, I was feeling pretty discouraged, when I happened upon a <a href="https://discourse.gohugo.io/t/shortcode-for-google-slide-presentation-embed/3464">Hugo forum post</a> with the HTML code someone used to embed a Google Slides presentation. I have hardly any familiarity with HTML, but clearly shortcodes and I weren’t getting along very well, so I decided to give it a shot. I copied and pasted the code from the forum, replacing the URL following <code>src=</code> with my link from above, and voila! My slides appeared on the post.</p>
<p>Here is the code for reference:</p>
<pre class="bash"><code>&lt;p&gt;&lt;iframe src=&quot;https://docs.google.com/presentation/d/e/2PACX-1vTsSj-oftLksGEKTkzCoBSIpzooO61ZMGAUzoGBv7u20OzMQK8ctX3gCpWTMm4UowrrWumFrdQsY0Cd/embed?start=false&amp;loop=false&amp;delayms=5000&quot; frameborder=&quot;0&quot; width=&quot;800&quot; height=&quot;600&quot; allowfullscreen=&quot;true&quot; mozallowfullscreen=&quot;true&quot; webkitallowfullscreen=&quot;true&quot;&gt;&lt;/iframe&gt;&lt;/p&gt;</code></pre>
</div>
</div>
<div id="moral-of-the-story" class="section level2">
<h2>Moral of the story</h2>
<p>It seems like shortcodes can be a really valuable feature of Hugo, especially should I ever need to include a video of some kind (or if I actually start using Twitter again <em>and</em> tweet something worth sharing, two improbable events). In the case of Google Docs/Sheets/etc., however, just using the HTML code snippet seems like the easiest option (and is a good motivation to learn HTML at a later date).</p>
</div>

    </div>
  
</div>

    </div>
  </section>

  
  
  

  

  

  

  

  

  
  

  
  
  

  
  
  
  
  

  
  

  <section id="2018-12-11-order-items" class="home-section wg-blank   "  >
    <div class="container">
      


<div class="row">
  
    <div class="col-12 col-lg-4 section-heading">
      <h1>Ordering website components</h1>
      
    </div>
    <div class="col-12 col-lg-8">
      


<p><img class="special-img-class" style="width:50%" src="/img/out_of_order.png" /></p>
<p>As I continue setting up this site, one of the major things irking me was the order of content on the website. For example, I wasn’t loving that the first thing you saw when visiting the site was the giant picture of my face. I wanted to be able to put content at the top of the site and move information like my bio and work experience towards the bottom. Hugo allows you to weight certain types of content to change the order, so that’s what I tackled.</p>
<div id="menu-links" class="section level2">
<h2>Menu links</h2>
<p>The first order of things I could change was the order of links in the menu on the top right of my page. In this case, I wanted my information to be the first link, followed by “Publications”, “Projects”, “Posts”, “Tutorials”, and “Contact”. The weighting for these links is found in the <code>config.toml</code> file of the website repo, where a standard entry for one of the links looks a bit like this:</p>
<pre class="bash"><code>[[menu.main]]
  name = &quot;Publications&quot;
  url = &quot;#publications&quot;
  weight = 2</code></pre>
<p>Here you can change the “name”, which is what the menu option actually says (e.g. I changed the menu option for my bio from “Home” to “About”). The URL refers to where on the site the link will point to. If you are trying to link to a homepage widget, as in the example above, the url will be <code>#</code> followed by the file name of that widget in the <code>content/home/</code> folder. You want to be careful not to link to a widget that is inactive (which you can see in the first few lines of the Markdown document for that widget), otherwise the link will not go anywhere.</p>
<p>If you are linking to something else, you can provide the file-path relative to the <code>content</code> folder – for example, my tutorials live in the folder <code>content/tutorial</code>:</p>
<pre class="bash"><code>[[menu.main]]
  name = &quot;Tutorials&quot;
  url = &quot;/tutorial/&quot;
  weight = 5</code></pre>
<p>Finally, you can change the weights to reflect the order you want the links to appear in the menu: the links will appear from left to right in ascending order of weight (<code>weight = 1</code> is the first link, etc.)</p>
</div>
<div id="home-page-content" class="section level2">
<h2>Home page content</h2>
<p>Initially, I confused changing the weights in the <code>config.toml</code> file with actually changing the order of those sections on the home page, but doing that actually requires editing the Markdown files for those types of content in the <code>content/home/</code> folder. For example, I decided I wanted to put posts first, so I went to <code>content/home/posts.md</code> and modified the <code>weight</code> argument so that it was lowest and continued editing the weights until everything was in the order I wanted.</p>
</div>

    </div>
  
</div>

    </div>
  </section>

  
  
  

  

  

  

  

  

  
  

  
  
  

  
  
  
  
  

  
  

  <section id="2018-12-7-create-website" class="home-section wg-blank   "  >
    <div class="container">
      


<div class="row">
  
    <div class="col-12 col-lg-4 section-heading">
      <h1>Creating this website</h1>
      
    </div>
    <div class="col-12 col-lg-8">
      


<p><img class="special-img-class" style="width:75%" src="/img/website_fig.png" /></p>
<div id="getting-started" class="section level2">
<h2>Getting started</h2>
<p>I’ve been contemplating the idea of creating a website for sometime, but my initial (mostly half-hearted) attempts ran into some serious technological glitches. I started out trying to create a GitHub site using Jekyll, since I knew I wanted my website to be hosted via GitHub to keep all of my code in the same place. After about an hour of unsuccessful commits that resulted in a lovely white page with my name on it, I scrapped the whole venture until I heard about the <a href="https://bookdown.org/yihui/blogdown/"><code>blogdown</code></a> R package. Since R is my programming language of choice, the idea of being able to write posts in RMarkdown and preview my site from the comfort of RStudio was tremendously appealing. I saw a research site created through Hugo’s Academic theme using <code>blogdown</code> that referenced some helpful tutorials and decided to give the process another shot.</p>
<div id="hugo-blogdown" class="section level3">
<h3>Hugo + <code>blogdown</code></h3>
<p>After my initial failures with Jekyll, I knew enough about the process (and my own skills) to be wary of the Academic theme’s optimistic tagline “Create a beautifully simple site in under 10 minutes”. So fair warning: getting this site up and semi-running was the work of an afternoon plus. I am not going to provide a full tutorial here, since several excellent ones already exist, but I will give my input on which resources worked for me and where I had difficulties.</p>
</div>
</div>
<div id="useful-resources" class="section level2">
<h2>Useful resources</h2>
<p>I started out following <a href="https://amber.rbind.io/blog/2016/12/19/creatingsite/">Amber Thomas’s instructions</a> for using <code>blogdown</code>, Hugo, and GitHub pages to create a website. I probably should have taken better note of her initial disclaimer that simpler methods of creating a site exist, but nonetheless I found her instructions useful for wrapping my head around the method behind the whole process. As she explains, GitHub pages and Hugo work in different ways: GitHub pages uses information from the master branch of the repository to build your website, while Hugo builds a website inside the public folder. The solution she presents in her tutorial creates a sub branch within GitHub. She provides some shell scripts written by <a href="https://www.hjdskes.nl">Jente Hidskes</a> that create the sub branch and also update the site. These did not run for me for whatever reason – I believe Hidskes has written an updated tutorial to fix them, however.</p>
<p>While I was reading the comments on Thomas’s tutorials to try to troubleshoot the shell script problems, I found a different <a href="http://robertmyles.github.io/2017/02/01/how-to-make-a-github-pages-blog-with-rstudio-and-hugo/">tutorial by Robert McDonnell</a> that promised a more streamlined approach. McDonnell gets around the GitHub/Hugo conflicts by creating two GitHub repositories, one for the website (for example, my repo is called <a href="https://github.com/babichmorrowc/babichmorrowc.github.io">babichmorrowc.github.io</a>) and one for all of the Hugo and <code>blogdown</code> content (mine is called <a href="https://github.com/babichmorrowc/website-hugo">website-hugo</a>).</p>
<p>To link these two repositories, Thomas has users create a <a href="https://blog.github.com/2016-02-01-working-with-submodules/">Git Submodule</a> using the following code:</p>
<pre class="bash"><code>git submodule add -b master git@github.com:&lt;username&gt;/&lt;username&gt;.github.io.git public</code></pre>
<p>This is where, once again, I ran into difficulties. I kept getting the error:</p>
<pre><code>Cloning into &#39;&lt;filepath&gt;/website-hugo/public&#39;...
git@github.com: Permission denied (publickey).
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
fatal: clone of &#39;git@github.com:babichmorrowc/babichmorrowc.github.io.git&#39; into submodule path &#39;&lt;filepath&gt;/website-hugo/public&#39; failed
</code></pre>
<p>Fortunately, there turned out to be an easy fix for this – the code McDonnell provides uses the SSH URL, which is something I do not have set up. I replaced it with the HTTPS URL (code below), and went on my merry way.</p>
<pre class="bash"><code>git submodule add -b master https://github.com/babichmorrowc/babichmorrowc.github.io.git public</code></pre>
</div>
<div id="pushing-to-github" class="section level2">
<h2>Pushing to GitHub</h2>
<p>From that point forward, I had a functioning site that I could preview in RStudio using the <code>serve_site()</code> function. Using the following Git commands, I could push those changes and get that same site up and running online:</p>
<pre class="bash"><code># From the folder for the repo website-hugo
cd public
git add .
git commit -m &quot;new commit!&quot;
git push origin master

cd ..
git add .
git commit -m &quot;new commit part 2&quot;
git push</code></pre>
</div>
<div id="editing-my-site" class="section level2">
<h2>Editing my site</h2>
<p>I’m still learning my way around the format of the Hugo site, but by some trial and error I’ve started the process of customizing the site. I am using the <a href="https://themes.gohugo.io/academic/">Academic</a> theme. So far I’ve managed to update my bio, add a publication, and take a stab at creating projects. As I get more familiarity with the format, I will try for a more comprehensive blog post explaining how to add your own content.</p>
</div>

    </div>
  
</div>

    </div>
  </section>



      

    
    
    
    <script src="/js/mathjax-config.js"></script>
    

    
    
    
      <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.imagesloaded/4.1.4/imagesloaded.pkgd.min.js" integrity="sha256-lqvxZrPLtfffUl2G/e7szqSvPBILGbwmsGE1MKlOi0Q=" crossorigin="anonymous"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.isotope/3.0.6/isotope.pkgd.min.js" integrity="sha256-CBrpuqrMhXwcLLUd5tvQ4euBHCdh7wGlDfNz8vbu/iI=" crossorigin="anonymous"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/fancybox/3.5.7/jquery.fancybox.min.js" integrity="sha256-yt2kYMy0w8AbtF89WXb2P1rfjcP/HTHLT7097U8Y5b8=" crossorigin="anonymous"></script>

      

      
        
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.15.10/highlight.min.js" integrity="sha256-1zu+3BnLYV9LdiY85uXMzii3bdrkelyp37e0ZyTAQh0=" crossorigin="anonymous"></script>
        
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.15.10/languages/r.min.js"></script>
        
      

      
      
      <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml.js" integrity="" crossorigin="anonymous" async></script>
      
    

    
    

    
    
    <script>hljs.initHighlightingOnLoad();</script>
    

    
    
    
    
    
    
    <script>
      const search_config = {"indexURI":"/index.json","minLength":1,"threshold":0.3};
      const i18n = {"no_results":"No results found","placeholder":"Search...","results":"results found"};
      const content_type = {
        'post': "Posts",
        'project': "Projects",
        'publication' : "Publications",
        'talk' : "Talks"
        };
    </script>
    

    
    

    
    
    <script id="search-hit-fuse-template" type="text/x-template">
      <div class="search-hit" id="summary-{{key}}">
      <div class="search-hit-content">
        <div class="search-hit-name">
          <a href="{{relpermalink}}">{{title}}</a>
          <div class="article-metadata search-hit-type">{{type}}</div>
          <p class="search-hit-description">{{snippet}}</p>
        </div>
      </div>
      </div>
    </script>
    

    
    
    <script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/3.2.1/fuse.min.js" integrity="sha256-VzgmKYmhsGNNN4Ph1kMW+BjoYJM2jV5i4IlFoeZA9XI=" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script>
    

    
    

    
    
    <script id="dsq-count-scr" src="https://babichmorrowc-github-io.disqus.com/count.js" async></script>
    

    
    
    
    
    
    
    
    
    
      
    
    
    
    
    <script src="/js/academic.min.d6bd04fdad2ad213aa8111c5a3b72fc5.js"></script>

    






  
  
  <div class="container">
    <footer class="site-footer">
  
  <p class="powered-by">
    
      <a href="/privacy/">Privacy Policy</a>
    
    
  </p>
  

  <p class="powered-by">
    2019 &middot; 

    Powered by the
    <a href="https://sourcethemes.com/academic/" target="_blank" rel="noopener">Academic theme</a> for
    <a href="https://gohugo.io" target="_blank" rel="noopener">Hugo</a>.

    
    <span class="float-right" aria-hidden="true">
      <a href="#" class="back-to-top">
        <span class="button_icon">
          <i class="fas fa-chevron-up fa-2x"></i>
        </span>
      </a>
    </span>
    
  </p>
</footer>

  </div>
  

  
<div id="modal" class="modal fade" role="dialog">
  <div class="modal-dialog">
    <div class="modal-content">
      <div class="modal-header">
        <h5 class="modal-title">Cite</h5>
        <button type="button" class="close" data-dismiss="modal" aria-label="Close">
          <span aria-hidden="true">&times;</span>
        </button>
      </div>
      <div class="modal-body">
        <pre><code class="tex hljs"></code></pre>
      </div>
      <div class="modal-footer">
        <a class="btn btn-outline-primary my-1 js-copy-cite" href="#" target="_blank">
          <i class="fas fa-copy"></i> Copy
        </a>
        <a class="btn btn-outline-primary my-1 js-download-cite" href="#" target="_blank">
          <i class="fas fa-download"></i> Download
        </a>
        <div id="modal-error"></div>
      </div>
    </div>
  </div>
</div>

</body>
</html>
